---
title: "4-normalization: auto + nabut"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Preparation

## Libraries
```{r}
library(SummarizedExperiment)
library(MultiAssayExperiment)
library(PCAtools)
library(ComplexHeatmap)
library(BatchQC)
library(sva)
library(tidyr)
library(dplyr)
library(tibble)
library(purrr)
library(ggplot2)
library(pheatmap)
library(ggpubr)
library(reshape2)
library(viridis)
library(ggrepel)
library(missForest)
library(limma)
library(stringr)
library(RColorBrewer)
library(scales)
```

## Functions
```{r}
source("~/single-cell-histone-ptm/functions.R")
```

## Load data
```{r}
# metadata
setwd("~/single-cell-histone-ptm/1-data/single-cell")

coldata <- read.csv("sample-table.csv", 
               header = TRUE,
               stringsAsFactors = TRUE)
# coldata[] <- lapply(coldata, as.factor) # make all vars factors

# skyline output
setwd("~/single-cell-histone-ptm/1-data/single-cell/intensities")

# combine all plates
df.1 <- read.csv("241113_Skyline_SC_export_auto1_cells.csv", 
               header = TRUE,
               stringsAsFactors = TRUE,
               na.strings = "#N/A")

df.2 <- read.csv("241106_Skyline_SC_export_auto2_auto4_cells.csv", 
               header = TRUE,
               stringsAsFactors = TRUE,
               na.strings = "#N/A")

df.3 <- read.csv("241111_Skyline_SC_export_nabut1_nabut4_cells.csv", 
               header = TRUE,
               stringsAsFactors = TRUE,
               na.strings = "#N/A")

df.combine <- rbind(df.1, df.2, df.3)

# filter peptides
df.combine.filter <- filter_peptides_skyline(df.combine)

# deconvolute isobaric H4 peptides using fragments
df.combine.filter.deconv <- deconvolute_h4(df.combine.filter)

# combine into summarized experiment object
obj <- load_summarized_experiment_skyline(df.combine.filter.deconv, coldata)
```

### Add derivitization efficiency data
```{r}
setwd("~/single-cell-histone-ptm/2-derivitization-efficiency/DIA")

prop.eff <- read.csv("propionylation_efficiency_percent.csv",
                     row.names = 1)

# order the same as obj
prop.eff <- prop.eff[match(obj$run_id, sub(".*_(\\d+)$", "\\1", rownames(prop.eff))),]
rownames(prop.eff) <- colnames(obj)
colData(obj) <- cbind(colData(obj), prop.eff)
```

### Modify colData
```{r}
# change replicate 4 to 3 to make it easier to read for auto
obj$replicate <- as.character(obj$replicate)
obj$replicate[obj$replicate == 4 & obj$experiment == "auto"] <- 3

obj$replicate[obj$experiment == "nabut"]

# change replicate 4 to 2 to make it easier to read for nabut
obj$replicate[obj$replicate == 4 & obj$experiment == "nabut"] <- 5
obj$replicate[obj$replicate == 1 & obj$experiment == "nabut"] <- 4
obj$replicate <- factor(obj$replicate)

# set cells from auto experiments to "control" condition - these were not treated
obj$condition[obj$experiment == "auto"] <- "control"

# combine experiment and replicate
obj$experiment_replicate <- factor(paste(obj$experiment, obj$replicate, sep = "_"))
```

### Save object
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut")

saveRDS(obj, "obj.RDS")
```

### Output tables
```{r}
dir.create("~/single-cell-histone-ptm/4-normalization/auto+nabut/unfiltered")
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/unfiltered")

write.csv(assay(obj, "MS1"), "MS1.csv")
write.csv(assay(obj, "MS2"), "MS2.csv")
write.csv(assay(obj, "RT"), "RT.csv")
write.csv(assay(obj, "RT")/40, "RT.relative.csv")
write.csv(rowData(obj)[,c("Protein.Accession",
                          "Protein.Gene",
                          "Peptide.Sequence",
                          "Precursor.Charge",
                          "PrecursorMz",
                          "Peptide.Note.Clean")],
          "peptidoform-annotation.csv")
```

# Unfiltered Data Charecterization

## Types of histone PTMs detected
```{r}
dir.create("~/single-cell-histone-ptm/4-normalization/auto+nabut/unfiltered")
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/unfiltered")

# Categorize each peptide
categories <- map_chr(rownames(obj), categorize_peptide)

# Count categories
category_counts <- table(categories)

# Convert to dataframe for ggplot
df <- as.data.frame(category_counts)
  

df$categories <- factor(df$categories, levels = c("Unmodified",
                                                  "Acetylated",
                                                  "Mono-methylated",
                                                  "Di-methylated",
                                                  "Tri-methylated",
                                                  "Hybrid modifications",
                                                  "Other modifications"))

pdf("PTM_types.pdf",
    width = 5,
    height = 5)
ggplot(df, aes(x = "", y = Freq, fill = categories)) +
  geom_bar(width = 1, stat = "identity", color = "white") +
  coord_polar("y", start = 0) +
  scale_fill_brewer(palette = "Set1") + # Use a color palette
  theme_void() +
  theme(legend.position = "right",  # Adjust legend position
        legend.title = element_text(size = 12),  # Adjust legend title size
        legend.text = element_text(size = 10)) +  # Adjust legend text size
  labs(fill = "Modification Type") +
  geom_text(aes(label = paste0(round(Freq / sum(Freq) * 100), "%")), 
            position = position_stack(vjust = 0.6), size = 4)# Add percentage labels
dev.off()
```

## Number of samples
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/unfiltered")

# by plate
df <- as.data.frame(table(obj$replicate))
names(df) <- c("Replicate", "Count")
pdf("samples_per_replicate.pdf",
    width = 5,
    height = 5)
ggplot(df, aes(x = Replicate, y = Count)) +
  geom_bar(stat = "identity") +
  coord_flip() +
  theme_Publication() +
  labs(title = "Number of samples per plate", x = "Plate Replicate", y = "Frequency") +
  scale_y_continuous(labels = scales::number_format(accuracy = 1))
dev.off()

# by type
df <- as.data.frame(table(obj$multiplier))
names(df) <- c("Replicate", "Count")
pdf("samples_per_type.pdf",
    width = 5,
    height = 5)
ggplot(df, aes(x = Replicate, y = Count)) +
  geom_bar(stat = "identity") +
  coord_flip() +
  theme_Publication() +
  labs(title = "Number of samples per type", x = "Amount of cells", y = "Frequency")
dev.off()
```

## Peptides detected
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/unfiltered")

# MS1 precursors
data <- assay(obj, "MS1")
data[data == 0] <- NA
df <- as.data.frame(colSums(!is.na(data)))
names(df) <- c("Count")

# number of missing values
sum(is.na(data))/length(!data)*100

pdf("MS1_precursors_per_sample.pdf",
    width = 5,
    height = 5)
ggplot(df, aes(x = Count)) +
  geom_histogram() +
  theme_Publication() +
  labs(title = "Number of precursors\ndetected per sample", x = "# Peptides", y = "Frequency")
dev.off()

pdf("MS1_missing_values.pdf",
    width = 5,
    height = 5)
set.seed(123)
pheatmap(data, 
         annotation_col = as.data.frame(colData(obj)[,c("replicate", "multiplier")]), 
         scale = "none", 
         na_col = "black", 
         show_rownames = FALSE,
         show_colnames = FALSE,
         cluster_rows = FALSE, 
         cluster_cols = FALSE)
dev.off()

# MS2 fragments
# not accurate because don't have data for individual fragments,
# just aggreagate for each peptide
data <- assay(obj, "MS2")
data[data == 0] <- NA
df <- as.data.frame(colSums(!is.na(data)))
names(df) <- c("Count")
pdf("MS2_fragments_per_sample.pdf",
    width = 5,
    height = 5)
ggplot(df, aes(x = Count)) +
  geom_histogram() +
  theme_Publication() +
  labs(title = "Number of fragments\ndetected per sample", x = "# Peptides", y = "Frequency")
dev.off()

pdf("MS2_missing_values.pdf",
    width = 5,
    height = 5)
set.seed(123)
data <- assay(obj, "MS2")
data[data == 0] <- NA
pheatmap(data, 
         annotation_col = as.data.frame(colData(obj)[,c("replicate", "multiplier")]), 
         scale = "none", 
         na_col = "black", 
         show_rownames = FALSE,
         show_colnames = FALSE,
         cluster_rows = FALSE, 
         cluster_cols = FALSE)
dev.off()
```

## Peptide retention times
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/unfiltered")

df <- assay(obj, "RT") %>% 
  rownames_to_column("id") %>%
  pivot_longer(
    cols = -id, # This keeps the 'id' column fixed
    names_to = "variable",
    values_to = "value") %>%
  merge(as.data.frame(colData(obj)), by.x = "variable", by.y = "row.names")

pdf("peptide_retention_times_unflitered_replicate.pdf",
    width = 10,
    height = 10)
ggplot(df, aes(x = id, y = value, color = as.factor(replicate))) +
  geom_jitter(alpha = 0.75) +
  theme_Publication() +
  labs(title = "", x = "Peptide", y = "Retention time (min)") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 10),
        legend.position = "bottom") +
  scale_color_discrete(name = "Replicate")
dev.off()

pdf("peptide_retention_times_unflitered_type.pdf",
    width = 10,
    height = 10)
ggplot(df, aes(x = id, y = value, color = as.factor(multiplier))) +
  geom_jitter(alpha = 0.75) +
  theme_Publication() +
  labs(title = "", x = "Peptide", y = "Retention time (min)") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 10),
        legend.position = "bottom") +
  scale_color_discrete(name = "Type")
dev.off()
```

## Intensity distribution

### MS1
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/unfiltered")

# transform
assays(obj)[["MS1.log2"]] <- log2(assay(obj, "MS1") + 1)

# plot by cell
df <- assay(obj, "MS1.log2") %>% 
  rownames_to_column("id") %>%
  pivot_longer(
    cols = -id, # This keeps the 'id' column fixed
    names_to = "variable",
    values_to = "value") %>%
  merge(as.data.frame(colData(obj)), by.x = "variable", by.y = "row.names")

pdf("MS1.log2_distribution_unflitered_cell.pdf",
    width = 8,
    height = 10)
ggplot(df, aes(x = as.factor(run_id), y = value, fill = as.factor(replicate))) +
  geom_boxplot(width = 0.7) +
  theme_Publication() +
  labs(title = "Cell #", x = "Run", y = "Log2(MS1 intensity)", fill = "Batch") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 10)) +
  facet_wrap(.~as.factor(multiplier), scales = "free_x", nrow = 3)
dev.off()

pdf("MS1.log2_distribution_unflitered_1cell_100cell.pdf",
    width = 7,
    height = 7)
ggplot(subset(df, multiplier == 1 | multiplier == 100), aes(x = as.factor(run_id), y = value, fill = as.factor(replicate))) +
  geom_boxplot(width = 0.7) +
  theme_Publication() +
  labs(x = "Injection", y = "Log2(MS1 intensity)", fill = "Batch") +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank()) + 
  facet_wrap(.~as.factor(multiplier), scales = "free_x", nrow =2)
dev.off()

# plot by peptide
pdf("MS1.log2_distribution_unflitered_peptide.pdf",
    width = 15,
    height = 10)
ggplot(df, aes(x = id, y = value)) +
  geom_boxplot() +
  theme_Publication() +
  labs(title = "Cell #", x = "Peptide", y = "Log2(MS1 intensity)", fill = "Batch") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 10)) +
  facet_wrap(.~as.factor(multiplier), scales = "free_x")
dev.off()

# heatmap
# column_ha = HeatmapAnnotation(replicate = as.factor(obj$replicate),
#                               cell_number = as.factor(obj$multiplier),
#                               run_id = obj$run_id)
# row_ha = rowAnnotation(charge = anno_barplot(rowData(obj)$Precursor.Charge))
# 
# pdf("MS1.log2_unflitered_heatmap.pdf",
#     width = 10,
#     height = 10)
# set.seed(123)
# Heatmap(scale(assay(obj, "MS1.log2"), center = FALSE, scale = FALSE),
#         top_annotation = column_ha,
#         right_annotation = row_ha,
#         heatmap_legend_param = list(title = "Log2(MS1 area)"),
#         cluster_rows = TRUE,
#         cluster_columns = TRUE,
#         show_column_names = FALSE,
#         row_names_gp = gpar(fontsize = 10))
# dev.off()

# PCA
df <- colData(obj)
df$multiplier <- factor(df$multiplier)
p <- pca(na.omit(assay(obj, "MS1.log2")),
         metadata = df,
         center = TRUE, 
         scale = TRUE)

screeplot(p)

pdf("MS1.log2_unflitered_eigencor.pdf",
    width = 5,
    height = 5)
eigencorplot(p,
             main = "Principle component correlations",
             cexMain = 1.5,
             metavars = c("replicate", "date", "run_id", "multiplier"),
             col = viridis(100),
             colCorval = 'firebrick',
             fontCorval = 2,
             cexCorval = 0.5,
             rotLabX = 45,
             posColKey = 'top')
dev.off()

pdf("MS1.log2_unflitered_replicate_pca.pdf",
    width = 5,
    height = 5)
biplot(p,
       colby = "replicate", 
       shape = "multiplier",
       legendPosition = 'right',
       hline = 0, 
       vline = 0,
       gridlines.major = FALSE, 
       gridlines.minor = FALSE,
       lab = NULL)
dev.off()
```

### MS2
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/unfiltered")

# transform
assays(obj)[["MS2.log2"]] <- log2(assay(obj, "MS2") + 1)

# plot by cell
df <- assay(obj, "MS2.log2") %>% 
  rownames_to_column("id") %>%
  pivot_longer(
    cols = -id, # This keeps the 'id' column fixed
    names_to = "variable",
    values_to = "value") %>%
  merge(as.data.frame(colData(obj)), by.x = "variable", by.y = "row.names")

pdf("MS2.log2_distribution_unflitered_cell.pdf",
    width = 8,
    height = 10)
ggplot(df, aes(x = as.factor(run_id), y = value, fill = as.factor(replicate))) +
  geom_boxplot(width = 0.7) +
  theme_Publication() +
  labs(title = "Cell #", x = "Run", y = "Log2(MS2 intensity)", fill = "Batch") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 10)) +
  facet_wrap(.~as.factor(multiplier), scales = "free_x", nrow = 3)
dev.off()

# plot by peptide
pdf("MS2.log2_distribution_unflitered_peptide.pdf",
    width = 15,
    height = 10)
ggplot(df, aes(x = id, y = value)) +
  geom_boxplot() +
  theme_Publication() +
  labs(title = "Cell #", x = "Peptide", y = "Log2(MS2 intensity)", fill = "Batch") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 10)) +
  facet_wrap(.~as.factor(multiplier), scales = "free_x")
dev.off()

# heatmap
# column_ha = HeatmapAnnotation(replicate = as.factor(obj$replicate),
#                               cell_number = as.factor(obj$multiplier),
#                               run_id = obj$run_id)
# row_ha = rowAnnotation(charge = anno_barplot(rowData(obj)$Precursor.Charge))
# 
# pdf("MS2.log2_unflitered_heatmap.pdf",
#     width = 10,
#     height = 10)
# set.seed(123)
# Heatmap(scale(assay(obj, "MS2.log2"), center = FALSE, scale = FALSE),
#         top_annotation = column_ha,
#         right_annotation = row_ha,
#         heatmap_legend_param = list(title = "Log2(MS2 area)"),
#         cluster_rows = TRUE,
#         cluster_columns = TRUE,
#         show_column_names = FALSE,
#         row_names_gp = gpar(fontsize = 10))
# dev.off()

# PCA
df <- colData(obj)
df$multiplier <- factor(df$multiplier)
p <- pca(na.omit(assay(obj, "MS2.log2")),
         metadata = df,
         center = TRUE, 
         scale = TRUE)

screeplot(p)

pdf("MS2.log2_unflitered_eigencor.pdf",
    width = 5,
    height = 5)
eigencorplot(p,
             main = "Principle component correlations",
             cexMain = 1.5,
             metavars = c("replicate", "date", "run_id", "multiplier"),
             col = viridis(100),
             colCorval = 'firebrick',
             fontCorval = 2,
             cexCorval = 0.5,
             rotLabX = 45,
             posColKey = 'top')
dev.off()

pdf("MS2.log2_unflitered_replicate_pca.pdf",
    width = 5,
    height = 5)
biplot(p,
       colby = "replicate", 
       shape = "multiplier",
       legendPosition = 'right',
       hline = 0, 
       vline = 0,
       gridlines.major = FALSE, 
       gridlines.minor = FALSE,
       lab = NULL)
dev.off()
```

### MS1 vs MS2
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/unfiltered")

# by standards
df.1 <- assay(obj, "MS1.log2") %>% 
  rownames_to_column("id") %>%
  pivot_longer(
    cols = -id, # This keeps the 'id' column fixed
    names_to = "variable",
    values_to = "value")


df.2 <- assay(obj, "MS2.log2") %>% 
  rownames_to_column("id") %>%
  pivot_longer(
    cols = -id, # This keeps the 'id' column fixed
    names_to = "variable",
    values_to = "value")

df <- cbind(df.1, df.2$value)
colnames(df) <- c("id", "variable", "ms1", "ms2")
df <- merge(df, as.data.frame(colData(obj)), by.x = "variable", by.y = "row.names")
df$id <- gsub(";", " ", df$id)

# by cell amount
pdf("MS1_vs_MS2_unflitered_type.pdf",
    width = 8,
    height = 8)
ggplot(subset(df, multiplier != "0"), aes(x = ms1, y = ms2, color = as.factor(replicate))) +
  geom_point() +
  stat_smooth(method = "lm") +
  stat_cor() +
  facet_wrap(.~as.factor(multiplier)) +
  labs(title = "", x = "Log2(MS1)", y = "Log2(MS2)") +
  theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_color_discrete(name = "Replicate")
dev.off()

# by peptide for single cells
pdf("MS1_vs_MS2_unflitered_peptide.pdf",
    width = 25,
    height = 25)
ggplot(subset(df, multiplier == "1"), aes(x = ms1, y = ms2, color = as.factor(replicate))) +
  geom_point() +
  stat_smooth(method = "lm") +
  stat_cor() +
  facet_wrap(.~as.factor(multiplier)) +
  labs(title = "HIstone Peptides", x = "Log2(MS1)", y = "Log2(MS2)") +
  theme_Publication(base_size = 12) +
  scale_y_continuous(labels = function(x) format(x, scientific = TRUE)) +
  facet_wrap(.~id, scales = "free", labeller = labeller(id = label_wrap_gen(20))) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_color_discrete(name = "Replicate")
dev.off()
```

## Background noise

### MS1
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/unfiltered")

# plot single cells vs blanks
df <- assay(obj, "MS1") %>% 
  rownames_to_column("id") %>%
  pivot_longer(
    cols = -id, # This keeps the 'id' column fixed
    names_to = "variable",
    values_to = "value") %>%
  merge(as.data.frame(colData(obj)), by.x = "variable", by.y = "row.names") %>%
  subset(multiplier %in% c(0, 1))
df$id <- gsub(";", " ", df$id)

# summarize
df %>%
  group_by(multiplier) %>%
  summarize(mean = mean(value, na.rm = TRUE))

# peptides separate
  # replicates merged
  pdf("MS1_0cell_vs_1cell_peptide_all.pdf",
    width = 25,
    height = 25)
  ggplot(df, aes(x = as.factor(multiplier), y = value, fill = as.factor(multiplier))) +
    geom_boxplot() +
    geom_jitter(alpha = 0.5) +
    scale_y_continuous(labels = function(x) format(x, scientific = TRUE)) +
    theme_Publication(base_size = 12) +
    labs(title = "", x = "Cell #", y = "MS1 Intensity") +
    facet_wrap(~id, scales = "free_y", labeller = labeller(id = label_wrap_gen(20))) +
    theme(legend.position = "none",
          axis.title.x = element_text(size = 32),
          axis.title.y = element_text(size = 32),
          axis.text = element_text(size = 16),
          legend.title = element_text(size = 32),
          legend.text = element_text(size = 32),
          strip.text = element_text(size = 12))
  dev.off()
  
  # replicates separated
  pdf("MS1_0cell_vs_1cell_peptide_replicate.pdf",
      width = 20,
      height = 20)
  ggplot(df, aes(x = as.factor(multiplier), y = value)) +
    geom_boxplot(outlier.shape = NA) +
    geom_jitter(alpha = 0.75, aes(color = as.factor(replicate))) +
    scale_y_continuous(labels = function(x) format(x, scientific = TRUE)) +
    theme_Publication(base_size = 12) +
    labs(title = "", x = "Cell #", y = "MS1 Intensity", color = "Replicate") +
    facet_wrap(~id, scales = "free_y", labeller = labeller(id = label_wrap_gen(20)))
  dev.off()
  
  # subset representative peptides
  rep_peptides <- c("H2A1B-K13[un] K15[ac]", "H31-K9[me2] K14[un]","H31-K18[un] K23[ac]", "H4-K5[ac] K8[un] K12[ac] K16[un]")
  facet_labels <- c("H2AK15Ac", "H3K9Me2", "H3K23Ac", "H4K5AcK12Ac")
  names(facet_labels) <- rep_peptides
  df.sub <- subset(df, id %in% rep_peptides)
  
  pdf("MS1_0cell_vs_1cell_peptide_representative.pdf",
      width = 7,
      height = 3.5)
  ggplot(df.sub, aes(x = as.factor(multiplier), y = value, fill = as.factor(replicate))) +
    geom_boxplot(outlier.shape = NA) +
    geom_point(alpha = 0.5, size = 0.5, 
               position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75), stat = "identity") +
    scale_y_continuous(labels = function(x) format(x, scientific = TRUE)) +
    scale_fill_discrete(name = "Replicate") +
    guides(fill = guide_legend(override.aes = list(shape = NA))) +
    labs(title = "", x = "Cell #", y = "MS1 Intensity") +
    facet_wrap(.~id, scales = "free", nrow = 1, labeller = labeller(id = facet_labels)) +
    theme_Publication(base_size = 12) +
    theme(legend.position = "top") +
    guides(color = guide_legend( 
      override.aes=list(shape = 19))) 
  dev.off()

# all peptides
  # replicates merged
  pdf("MS1.log2_0cell_vs_1cell_all.pdf",
      width = 3,
      height = 3)
  ggplot(df, aes(x = value, fill = as.factor(multiplier))) +
    geom_density(alpha = 0.5) +
    scale_x_log10() +
    theme_Publication(base_size = 12) +
    labs(title = "", x = "Log2(MS1 intensity)", y = "Density", fill = "Cell #")
  dev.off()
  
  # replicates separated
  pdf("MS1.log2_0cell_vs_1cell_replicate.pdf",
      width = 5,
      height = 3)
  ggplot(df, aes(x = value, fill = as.factor(multiplier))) +
    geom_density(alpha = 0.5) +
    scale_x_log10() +
    theme_Publication(base_size = 12) +
    labs(title = "", x = "Log2(MS1 intensity)", y = "Density", fill = "Cell #") +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    facet_wrap(~replicate, nrow = 1)
  dev.off()
```

### MS2
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/unfiltered")

# plot single cells vs blanks
df <- assay(obj, "MS2") %>% 
  rownames_to_column("id") %>%
  pivot_longer(
    cols = -id, # This keeps the 'id' column fixed
    names_to = "variable",
    values_to = "value") %>%
  merge(as.data.frame(colData(obj)), by.x = "variable", by.y = "row.names") %>%
  subset(multiplier %in% c(0, 1))
df$id <- gsub(";", " ", df$id)

# summarize
df %>%
  group_by(multiplier) %>%
  summarize(mean = mean(value, na.rm = TRUE))

# peptides separate
  # replicates merged
  pdf("MS2_0cell_vs_1cell_peptide_all.pdf",
      width = 20,
      height = 20)
  ggplot(df, aes(x = as.factor(multiplier), y = value, fill = as.factor(multiplier))) +
    geom_boxplot() +
    geom_jitter(alpha = 0.5) +
    scale_y_continuous(labels = function(x) format(x, scientific = TRUE)) +
    theme_Publication(base_size = 12) +
    labs(title = "", x = "Cell #", y = "MS2 Intensity") +
    facet_wrap(~id, scales = "free_y", labeller = labeller(id = label_wrap_gen(20))) +
    theme(legend.position = "none")
  dev.off()
  
  # replicates separated
  pdf("MS2_0cell_vs_1cell_peptide_replicate.pdf",
      width = 20,
      height = 20)
  ggplot(df, aes(x = as.factor(multiplier), y = value)) +
    geom_boxplot(outlier.shape = NA) +
    geom_jitter(alpha = 0.75, aes(color = as.factor(replicate))) +
    scale_y_continuous(labels = function(x) format(x, scientific = TRUE)) +
    theme_Publication(base_size = 12) +
    labs(title = "", x = "Cell #", y = "MS2 Intensity", color = "Replicate") +
    facet_wrap(~id, scales = "free_y", labeller = labeller(id = label_wrap_gen(20)))
  dev.off()
  
    # subset representative peptides
  rep_peptides <- c("H2A1B-K13[un] K15[ac]", "H31-K9[me2] K14[un]","H31-K18[un] K23[ac]", "H4-K5[ac] K8[un] K12[ac] K16[un]")
  facet_labels <- c("H2AK15Ac", "H3K9Me2", "H3K23Ac", "H4K5AcK12Ac")
  names(facet_labels) <- rep_peptides
  df.sub <- subset(df, id %in% rep_peptides)
  
  pdf("MS2_0cell_vs_1cell_peptide_representative.pdf",
      width = 7,
      height = 3.5)
  ggplot(df.sub, aes(x = as.factor(multiplier), y = value, fill = as.factor(replicate))) +
    geom_boxplot(outlier.shape = NA) +
    geom_point(alpha = 0.5, size = 0.5, 
               position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75), stat = "identity") +
    scale_y_continuous(labels = function(x) format(x, scientific = TRUE)) +
    scale_fill_discrete(name = "Replicate") +
    guides(fill = guide_legend(override.aes = list(shape = NA))) +
    labs(title = "", x = "Cell #", y = "MS2 Intensity") +
    facet_wrap(.~id, scales = "free", nrow = 1, labeller = labeller(id = facet_labels)) +
    theme_Publication(base_size = 12) +
    theme(legend.position = "top") +
    guides(color = guide_legend( 
      override.aes=list(shape = 19))) 
  dev.off()


# all peptides
  # replicates merged
  pdf("MS2.log2_0cell_vs_1cell_all.pdf",
      width = 3,
      height = 3)
  ggplot(df, aes(x = value, fill = as.factor(multiplier))) +
    geom_density(alpha = 0.5) +
    scale_x_log10() +
    theme_Publication(base_size = 12) +
    labs(title = "", x = "Log2(MS2 intensity)", y = "Density", fill = "Cell #")
  dev.off()
  
  # replicates separated
  pdf("MS2.log2_0cell_vs_1cell_replicate.pdf",
      width = 5,
      height = 3)
  ggplot(df, aes(x = value, fill = as.factor(multiplier))) +
    geom_density(alpha = 0.5) +
    scale_x_log10() +
    theme_Publication(base_size = 12) +
    labs(title = "", x = "Log2(MS2 intensity)", y = "Density", fill = "Cell #") +
    facet_wrap(~replicate, nrow = 1) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
  dev.off()
```

## Derivitization efficiency
```{r}
setwd("/Users/ronaldcutler/Dropbox\ (EinsteinMed)/Sidoli-lab/Projects/Single-cell-histone/Analysis/4-normalization/auto+nabut/unfiltered")

df <- colData(obj) %>% 
  as.data.frame() %>%
  rownames_to_column("run") %>%
  pivot_longer(
    cols = starts_with("prop"),
    names_to = "variable",
    values_to = "value") %>%
  filter(multiplier == 1 | multiplier == 100)
df$variable <- factor(df$variable,
                      levels = c("prop.eff.underprop.summary",
                                 "prop.eff.overprop",
                                 "prop.eff.underprop.first.round",
                                 "prop.eff.underprop.second.round",
                                 "prop.eff.underprop.both.round"))

# summary
pdf("propionylation_efficiency.pdf",
    width = 8,
    height = 5)
ggplot(subset(df, replicate == 4 | replicate == 5), aes(x = variable, y = value)) +
  geom_boxplot(aes(fill = replicate), outlier.shape = NA, position = position_dodge(width = 0.75)) +
  geom_jitter(aes(fill = replicate, shape = as.factor(multiplier)), 
              position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75, jitter.height = 0), 
              alpha = 0.75, size = 1.5) +
  scale_fill_manual(values = brewer.pal(5, "Set1")[4:5]) + 
  scale_y_continuous(labels = label_percent(accuracy = 1)) +
  scale_x_discrete(labels = c("Under\nPropionylation", 
                              "Over\nPropionlyation", 
                              "Under\nPropionylation\n1st Round", 
                              "Under\nPropionylation\n2nd Round", 
                              "Under\nPropionylation\n1st & 2nd Rounds")) +
  labs(x = "", 
       shape = "# Cells Per Injection",
       y = "% Raw MS1 Area of Peptide\nNot Correctly Propionylated", 
       fill = "Batch") +
  theme_Publication() +
  theme(legend.box = "vertical",
        axis.text.x = element_text(angle = 45, vjust = 0.75, hjust=0.5))
dev.off()

pdf("propionylation_efficiency_summary.pdf",
    width = 5,
    height = 5)
ggplot(subset(df, !grepl("round", variable) & (replicate == 4 | replicate == 5)), aes(x = variable, y = value)) +
   geom_boxplot(aes(fill = replicate), outlier.shape = NA, position = position_dodge(width = 0.75)) +
  geom_jitter(aes(fill = replicate, shape = as.factor(multiplier)), 
              position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75, jitter.height = 0), 
              alpha = 0.75, size = 1.5) +
  scale_fill_manual(values = brewer.pal(5, "Set1")[4:5]) + 
  scale_y_continuous(labels = label_percent(accuracy = 1)) +
  scale_x_discrete(labels = c("Under\nPropionylation", "Over\nPropionlyation")) +
  labs(x = "", 
       shape = "# Cells Per Injection",
       y = "% Raw MS1 Area of Peptide\nNot Correctly Propionylated", 
       fill = "Batch") +
  theme_Publication() +
  theme(legend.box = "vertical")
dev.off()
```

### Correlation to abundance
- Propionlyation efficiency was estimated using H31-K9[un];K14[un]
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/unfiltered")

# correlation of abundance and prop eff
df <- assay(obj, "MS1") %>% 
  rownames_to_column("id") %>%
  pivot_longer(
    cols = -id, # This keeps the 'id' column fixed
    names_to = "variable",
    values_to = "value") %>%
  merge(as.data.frame(colData(obj)), by.x = "variable", by.y = "row.names") %>%
  filter(id == "H31-K9[un];K14[un]")

pdf("MS1.log1p_H31-K9[un];K14[un]_vs_prop.eff.underprop.summary.pdf",
    width = 5,
    height = 5)
ggplot(df, aes(x = log1p(value), y = prop.eff.underprop.summary)) +
  geom_point(aes(color = replicate), size = 1.5, alpha = 0.75) +
  stat_smooth(method = "lm") +
  stat_cor() +
  labs(x = "H31-K9[un];K14[un]\n  Raw Log1p Intensity",
       y = "% Raw MS1 Area of Peptide\nUnder Propionylated",
       color = "Batch") +
  scale_y_continuous(labels = label_percent(accuracy = 1)) +
  theme_Publication() +
  theme(legend.position = "right",       # Position legend on the right
        legend.direction = "vertical")   # Arrange legend items vertically
dev.off()

pdf("MS1.log1p_H31-K9[un];K14[un]_vs_prop.eff.overprop.pdf",
    width = 5,
    height = 5)
ggplot(df, aes(x = log1p(value), y = prop.eff.overprop)) +
  geom_point(aes(color = replicate), size = 1.5, alpha = 0.75) +
  stat_smooth(method = "lm") +
  stat_cor() +
  labs(x = "H31-K9[un];K14[un]\n  Raw Log1p Intensity",
       y = "% Raw MS1 Area of Peptide\nOver Propionylated",
       color = "Batch") +
  scale_y_continuous(labels = label_percent(accuracy = 1)) +
  theme_Publication() +
  theme(legend.position = "right",       # Position legend on the right
        legend.direction = "vertical")   # Arrange legend items vertically
dev.off()
```

# Filtering

## Cell level

### Low overall signal intensity
- Mean intensity of cell is 1 MAD away from that of all cells
- MAD threshold is set so that empty cells are filtered out - so any cells that are similar to these
```{r}
dir.create("~/single-cell-histone-ptm/4-normalization/auto+nabut/filtered")
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/filtered")

# filter out 2, 4, 8 cells
obj.cell_filter_1 <- obj[,obj$multiplier == 0 |
                     obj$multiplier == 1 |
                     obj$multiplier == 100]

# empty should be filtered
obj.cell_filter_2 <- filter_overall_intensity(obj.cell_filter_1, assay = "MS1", dev = 1, filter_low = TRUE, filter_high = FALSE)

# change coldata
obj.cell_filter_2$multiplier <- factor(obj.cell_filter_2$multiplier)

saveRDS(obj.cell_filter_2, "obj.cell_filter.RDS")
```

### Low propionylation efficiency
- Propiopnlyation efficiency of cell is 1 MAD away from that of all cells
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/filtered")

obj.underprop_eff_filter <- filter_prop_eff(obj.cell_filter_2, 
                                            prop.eff = obj.cell_filter_2$prop.eff.underprop.summary, 
                                            dev = 2, 
                                            filter_low = FALSE, 
                                            filter_high = TRUE)

obj.overprop_eff_filter <- filter_prop_eff(obj.underprop_eff_filter, 
                                          prop.eff = obj.underprop_eff_filter$prop.eff.overprop, 
                                          dev = 2, 
                                          filter_low = TRUE, 
                                          filter_high = TRUE)

saveRDS(obj.underprop_eff_filter, "obj.prop_eff_filter.RDS")
```

## Peptidoform level

### Inconsistent detection
- Filter out peptides not detected in > 50% of single-cells
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/filtered")

obj.peptide_filter <- filter_missing_peptides(obj.underprop_eff_filter, threshold = 0.5)

saveRDS(obj.peptide_filter, "obj.peptide_filter.RDS")
```

### Low signal
- Filter out peptides whose single cell signal is less than 10% greater than the background signal
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/filtered")

# add back empty samples temporarily
obj.peptide_filter_empty <- cbind(obj.peptide_filter, 
                                  obj.cell_filter_1[rownames(obj.peptide_filter), obj.cell_filter_1$multiplier == 0])

obj.background_filter <- filter_background_peptides(obj.peptide_filter_empty, ratio = 1.1)

saveRDS(obj.background_filter, "obj.background_filter.RDS")
```

#### Plots
- Plot now because empty cells will be removed in next filter

##### MS1
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/filtered")

# plot single cells vs blanks
df <- assay(obj.background_filter, "MS1") %>% 
  rownames_to_column("id") %>%
  pivot_longer(
    cols = -id, # This keeps the 'id' column fixed
    names_to = "variable",
    values_to = "value") %>%
  merge(as.data.frame(colData(obj.background_filter)), by.x = "variable", by.y = "row.names") %>%
  subset(multiplier %in% c(0, 1))
df$id <- gsub(";", " ", df$id)
df$multiplier <- factor(df$multiplier, levels = c(0, 1))

# peptides separate
  # replicates merged
  pdf("MS1_0cell_vs_1cell_peptide_all.pdf",
      width = 20,
      height = 20)
  ggplot(df, aes(x = as.factor(multiplier), y = value, fill = as.factor(multiplier))) +
    geom_boxplot() +
    geom_jitter(alpha = 0.5) +
    scale_y_continuous(labels = function(x) format(x, scientific = TRUE)) +
    theme_Publication(base_size = 12) +
    labs(title = "", x = "Cell #", y = "MS1 Intensity") +
    facet_wrap(~id, scales = "free_y", labeller = labeller(id = label_wrap_gen(20))) +
        theme(legend.position = "none",
          axis.title.x = element_text(size = 32),
          axis.title.y = element_text(size = 32),
          axis.text = element_text(size = 16),
          legend.title = element_text(size = 32),
          legend.text = element_text(size = 32),
          strip.text = element_text(size = 12))
  dev.off()
  
  # replicates separated
  pdf("MS1_0cell_vs_1cell_peptide_replicate.pdf",
      width = 20,
      height = 20)
  ggplot(df, aes(x = as.factor(multiplier), y = value)) +
    geom_boxplot(outlier.shape = NA) +
    geom_jitter(alpha = 0.75, aes(color = as.factor(replicate))) +
    scale_y_continuous(labels = function(x) format(x, scientific = TRUE)) +
    theme_Publication(base_size = 12) +
    labs(title = "", x = "Cell #", y = "MS1 Intensity", color = "Replicate") +
    facet_wrap(~id, scales = "free_y", labeller = labeller(id = label_wrap_gen(20)))
  dev.off()
  
  # subset representative peptides
  rep_peptides <- c("H2A1B-K13[un] K15[ac]", "H31-K9[me2] K14[un]","H31-K18[un] K23[ac]", "H4-K5[ac] K8[un] K12[ac] K16[un]")
  facet_labels <- c("H2AK15Ac", "H3K9Me2", "H3K23Ac", "H4K5AcK12Ac")
  names(facet_labels) <- rep_peptides
  df.sub <- subset(df, id %in% rep_peptides)
  
  pdf("MS1_0cell_vs_1cell_peptide_representative.pdf",
      width = 7,
      height = 3.5)
  ggplot(df.sub, aes(x = as.factor(multiplier), y = value, fill = as.factor(replicate))) +
    geom_boxplot(outlier.shape = NA) +
    geom_point(alpha = 0.5, size = 0.5, 
               position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75), stat = "identity") +
    scale_y_continuous(labels = function(x) format(x, scientific = TRUE)) +
    scale_fill_discrete(name = "Replicate") +
    guides(fill = guide_legend(override.aes = list(shape = NA))) +
    labs(title = "", x = "Cell #", y = "MS1 Intensity") +
    facet_wrap(.~id, scales = "free", nrow = 1, labeller = labeller(id = facet_labels)) +
    theme_Publication(base_size = 12) +
    theme(legend.position = "top") +
    guides(color = guide_legend( 
      override.aes=list(shape = 19))) 
  dev.off()

# all peptides
  # replicates merged
  pdf("MS1.log2_0cell_vs_1cell_all.pdf",
      width = 3,
      height = 3)
  ggplot(df, aes(x = value, fill = as.factor(multiplier))) +
    geom_density(alpha = 0.5) +
    scale_x_log10() +
    theme_Publication(base_size = 12) +
    labs(title = "", x = "Log2(MS1 intensity)", y = "Density", fill = "Cell #")
  dev.off()
  
  # replicates separated
  pdf("MS1.log2_0cell_vs_1cell_replicate.pdf",
      width = 5,
      height = 3)
  ggplot(df, aes(x = value, fill = as.factor(multiplier))) +
    geom_density(alpha = 0.5) +
    scale_x_log10() +
    theme_Publication(base_size = 12) +
    labs(title = "", x = "Log2(MS1 intensity)", y = "Density", fill = "Cell #") +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    facet_wrap(~replicate, nrow = 1)
  dev.off()
```

##### MS2
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/filtered")

# plot single cells vs blanks
df <- assay(obj.background_filter, "MS2") %>% 
  rownames_to_column("id") %>%
  pivot_longer(
    cols = -id, # This keeps the 'id' column fixed
    names_to = "variable",
    values_to = "value") %>%
  merge(as.data.frame(colData(obj.background_filter)), by.x = "variable", by.y = "row.names") %>%
  subset(multiplier %in% c(0, 1))
df$id <- gsub(";", " ", df$id)

# peptides separate
  # replicates merged
  pdf("MS2_0cell_vs_1cell_peptide_all.pdf",
      width = 20,
      height = 20)
  ggplot(df, aes(x = as.factor(multiplier), y = value, fill = as.factor(multiplier))) +
    geom_boxplot() +
    geom_jitter(alpha = 0.5) +
    scale_y_continuous(labels = function(x) format(x, scientific = TRUE)) +
    theme_Publication(base_size = 12) +
    labs(title = "", x = "Cell #", y = "MS2 Intensity") +
    facet_wrap(~id, scales = "free_y", labeller = labeller(id = label_wrap_gen(20))) +
    theme(legend.position = "none")
  dev.off()
  
  # replicates separated
  pdf("MS2_0cell_vs_1cell_peptide_replicate.pdf",
      width = 20,
      height = 20)
  ggplot(df, aes(x = as.factor(multiplier), y = value)) +
    geom_boxplot(outlier.shape = NA) +
    geom_jitter(alpha = 0.75, aes(color = as.factor(replicate))) +
    scale_y_continuous(labels = function(x) format(x, scientific = TRUE)) +
    theme_Publication(base_size = 12) +
    labs(title = "", x = "Cell #", y = "MS2 Intensity", color = "Replicate") +
    facet_wrap(~id, scales = "free_y", labeller = labeller(id = label_wrap_gen(20)))
  dev.off()
  
    # subset representative peptides
  rep_peptides <- c("H2A1B-K13[un] K15[ac]", "H31-K9[me2] K14[un]","H31-K18[un] K23[ac]", "H4-K5[ac] K8[un] K12[ac] K16[un]")
  facet_labels <- c("H2AK15Ac", "H3K9Me2", "H3K23Ac", "H4K5AcK12Ac")
  names(facet_labels) <- rep_peptides
  df.sub <- subset(df, id %in% rep_peptides)
  
  pdf("MS2_0cell_vs_1cell_peptide_representative.pdf",
      width = 7,
      height = 3.5)
  ggplot(df.sub, aes(x = as.factor(multiplier), y = value, fill = as.factor(replicate))) +
    geom_boxplot(outlier.shape = NA) +
    geom_point(alpha = 0.5, size = 0.5, 
               position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75), stat = "identity") +
    scale_y_continuous(labels = function(x) format(x, scientific = TRUE)) +
    scale_fill_discrete(name = "Replicate") +
    guides(fill = guide_legend(override.aes = list(shape = NA))) +
    labs(title = "", x = "Cell #", y = "MS2 Intensity") +
    facet_wrap(.~id, scales = "free", nrow = 1, labeller = labeller(id = facet_labels)) +
    theme_Publication(base_size = 12) +
    theme(legend.position = "top") +
    guides(color = guide_legend( 
      override.aes=list(shape = 19))) 
  dev.off()


# all peptides
  # replicates merged
  pdf("MS2.log2_0cell_vs_1cell_all.pdf",
      width = 3,
      height = 3)
  ggplot(df, aes(x = value, fill = as.factor(multiplier))) +
    geom_density(alpha = 0.5) +
    scale_x_log10() +
    theme_Publication(base_size = 12) +
    labs(title = "", x = "Log2(MS2 intensity)", y = "Density", fill = "Cell #")
  dev.off()
  
  # replicates separated
  pdf("MS2.log2_0cell_vs_1cell_replicate.pdf",
      width = 5,
      height = 3)
  ggplot(df, aes(x = value, fill = as.factor(multiplier))) +
    geom_density(alpha = 0.5) +
    scale_x_log10() +
    theme_Publication(base_size = 12) +
    labs(title = "", x = "Log2(MS2 intensity)", y = "Density", fill = "Cell #") +
    facet_wrap(~replicate, nrow = 1) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
  dev.off()
```


### Poor quantification
- Peptidoforms which do not have the expected correlation based on titration curves
```{r}
setwd("~/single-cell-histone-ptm/3-calibration-curves/auto/cells/unfiltered")

quant_filter <- read.csv("MS1_calibration_curve_correlations.csv")
 
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/filtered")

obj.quant_filter <- filter_non_linear_peptides(quant_filter, obj.background_filter, correlation = 0.8, p_value = 0.05)

saveRDS(obj.quant_filter, "obj.quant_filter.RDS")
```

## Empty injections
```{r}
obj.filter <- obj.quant_filter[,obj.quant_filter$multiplier != 0]
```

## Save object
- re-naming and saving in main folder
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut")

saveRDS(obj.filter, "obj.filter.RDS")
```

### Output tables
```{r}
dir.create("/Users/ronaldcutler/Dropbox\ (EinsteinMed)/Sidoli-lab/Projects/Single-cell-histone/Analysis/4-normalization/auto/unfiltered")
setwd("/Users/ronaldcutler/Dropbox\ (EinsteinMed)/Sidoli-lab/Projects/Single-cell-histone/Analysis/4-normalization/auto/unfiltered")

write.csv(assay(obj, "MS1"), "MS1.csv")
write.csv(assay(obj, "MS2"), "MS2.csv")
write.csv(assay(obj, "RT"), "RT.csv")
write.csv(assay(obj, "RT")/40, "RT.relative.csv")
write.csv(rowData(obj)[,c("Protein.Accession",
                          "Protein.Gene",
                          "Peptide.Sequence",
                          "Precursor.Charge",
                          "PrecursorMz",
                          "Peptide.Note.Clean")],
          "peptidoform-annotation.csv")
```

# Filtered Data Charecterization

## Types of histone PTMs detected
```{r}
dir.create("~/single-cell-histone-ptm/4-normalization/auto+nabut/filtered")
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/filtered")

# Categorize each peptide
categories <- map_chr(rownames(obj.filter), categorize_peptide)

# Count categories
category_counts <- table(categories)

# Convert to dataframe for ggplot
df <- as.data.frame(category_counts)
  

df$categories <- factor(df$categories, levels = c("Unmodified",
                                                  "Acetylated",
                                                  "Mono-methylated",
                                                  "Di-methylated",
                                                  "Tri-methylated",
                                                  "Hybrid modifications",
                                                  "Other modifications"))

pdf("PTM_types.pdf",
    width = 5,
    height = 5)
ggplot(df, aes(x = "", y = Freq, fill = categories)) +
  geom_bar(width = 1, stat = "identity", color = "white") +
  coord_polar("y", start = 0) +
  scale_fill_brewer(palette = "Set1") + # Use a color palette
  theme_void() +
  theme(legend.position = "right",  # Adjust legend position
        legend.title = element_text(size = 12),  # Adjust legend title size
        legend.text = element_text(size = 10)) +  # Adjust legend text size
  labs(fill = "Modification Type") +
  geom_text(aes(label = paste0(round(Freq / sum(Freq) * 100), "%")), 
            position = position_stack(vjust = 0.6), size = 4)# Add percentage labels
dev.off()
```

## Number of samples
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/filtered")

# by plate
df <- as.data.frame(table(obj.filter$replicate))
names(df) <- c("Replicate", "Count")
pdf("samples_per_replicate.pdf",
    width = 5,
    height = 5)
ggplot(df, aes(x = Replicate, y = Count)) +
  geom_bar(stat = "identity") +
  coord_flip() +
  theme_Publication() +
  labs(title = "Number of samples per plate", x = "Plate Replicate", y = "Frequency") +
  scale_y_continuous(labels = scales::number_format(accuracy = 1))
dev.off()

# by type
df <- as.data.frame(table(obj.filter$multiplier))
names(df) <- c("Replicate", "Count")
pdf("samples_per_type.pdf",
    width = 5,
    height = 5)
ggplot(df, aes(x = Replicate, y = Count)) +
  geom_bar(stat = "identity") +
  coord_flip() +
  theme_Publication() +
  labs(title = "Number of samples per type", x = "Amount of cells", y = "Frequency")
dev.off()
```

## Peptides detected
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/filtered")

# MS1 precursors
data <- assay(obj.filter, "MS1")
data[data == 0] <- NA
df <- as.data.frame(colSums(!is.na(data)))
names(df) <- c("Count")

# number of missing values
sum(is.na(data))/length(!data)*100

pdf("MS1_precursors_per_sample.pdf",
    width = 5,
    height = 5)
ggplot(df, aes(x = Count)) +
  geom_histogram() +
  theme_Publication() +
  labs(title = "Number of precursors\ndetected per sample", x = "# Peptides", y = "Frequency")
dev.off()

pdf("MS1_missing_values.pdf",
    width = 5,
    height = 5)
set.seed(123)
pheatmap(data, 
         annotation_col = as.data.frame(colData(obj.filter)[,c("replicate", "multiplier")]), 
         scale = "none", 
         na_col = "black", 
         show_rownames = FALSE,
         show_colnames = FALSE,
         cluster_rows = FALSE, 
         cluster_cols = FALSE)
dev.off()

# MS2 fragments
# not accurate because don't have data for individual fragments,
# just aggreagate for each peptide
data <- assay(obj.filter, "MS2")
data[data == 0] <- NA
df <- as.data.frame(colSums(!is.na(data)))
names(df) <- c("Count")
pdf("MS2_fragments_per_sample.pdf",
    width = 5,
    height = 5)
ggplot(df, aes(x = Count)) +
  geom_histogram() +
  theme_Publication() +
  labs(title = "Number of fragments\ndetected per sample", x = "# Peptides", y = "Frequency")
dev.off()

pdf("MS2_missing_values.pdf",
    width = 5,
    height = 5)
set.seed(123)
data <- assay(obj.filter, "MS2")
data[data == 0] <- NA
pheatmap(data, 
         annotation_col = as.data.frame(colData(obj.filter)[,c("replicate", "multiplier")]), 
         scale = "none", 
         na_col = "black", 
         show_rownames = FALSE,
         show_colnames = FALSE,
         cluster_rows = FALSE, 
         cluster_cols = FALSE)
dev.off()
```

## Peptide retention times
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/filtered")

df <- assay(obj.filter, "RT") %>% 
  rownames_to_column("id") %>%
  pivot_longer(
    cols = -id, # This keeps the 'id' column fixed
    names_to = "variable",
    values_to = "value") %>%
  merge(as.data.frame(colData(obj.filter)), by.x = "variable", by.y = "row.names")

pdf("peptide_retention_times_filtered_replicate.pdf",
    width = 10,
    height = 10)
ggplot(df, aes(x = id, y = value, color = as.factor(replicate))) +
  geom_jitter(alpha = 0.75) +
  theme_Publication() +
  labs(title = "", x = "Peptide", y = "Retention time (min)") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 10),
        legend.position = "bottom") +
  scale_color_discrete(name = "Replicate")
dev.off()

pdf("peptide_retention_times_filtered_type.pdf",
    width = 10,
    height = 10)
ggplot(df, aes(x = id, y = value, color = as.factor(multiplier))) +
  geom_jitter(alpha = 0.75) +
  theme_Publication() +
  labs(title = "", x = "Peptide", y = "Retention time (min)") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 10),
        legend.position = "bottom") +
  scale_color_discrete(name = "Type")
dev.off()

# single cells only
pdf("peptide_retention_times_filtered_1cell.pdf",
    width = 6,
    height = 15)
ggplot(subset(df, multiplier == "1"), aes(x = id, y = value, color = as.factor(replicate))) +
  geom_jitter(alpha = 0.75) +
  coord_flip() +
  theme_Publication() +
  labs(title = "", x = "Peptide", y = "Retention time (min)") +
  theme(legend.position = "bottom") +
  scale_color_discrete(name = "Replicate")
dev.off()
```

## Derivitization efficiency
```{r}
setwd("/Users/ronaldcutler/Dropbox\ (EinsteinMed)/Sidoli-lab/Projects/Single-cell-histone/Analysis/4-normalization/auto+nabut/filtered")

df <- colData(obj.filter) %>% 
  as.data.frame() %>%
  rownames_to_column("run") %>%
  pivot_longer(
    cols = starts_with("prop"),
    names_to = "variable",
    values_to = "value") %>%
  filter(multiplier == 1 | multiplier == 100)
df$variable <- factor(df$variable,
                      levels = c("prop.eff.underprop.summary",
                                 "prop.eff.overprop",
                                 "prop.eff.underprop.first.round",
                                 "prop.eff.underprop.second.round",
                                 "prop.eff.underprop.both.round"))

# summary
pdf("propionylation_efficiency.pdf",
    width = 8,
    height = 5)
ggplot(subset(df, replicate == 4 | replicate == 5), aes(x = variable, y = value)) +
  geom_boxplot(aes(fill = replicate), outlier.shape = NA, position = position_dodge(width = 0.75)) +
  geom_jitter(aes(fill = replicate, shape = as.factor(multiplier)), 
              position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75, jitter.height = 0), 
              alpha = 0.75, size = 1.5) +
  scale_fill_manual(values = brewer.pal(5, "Set1")[4:5]) + 
  scale_y_continuous(labels = label_percent(accuracy = 1)) +
  scale_x_discrete(labels = c("Under\nPropionylation", 
                              "Over\nPropionlyation", 
                              "Under\nPropionylation\n1st Round", 
                              "Under\nPropionylation\n2nd Round", 
                              "Under\nPropionylation\n1st & 2nd Rounds")) +
  labs(x = "", 
       shape = "# Cells Per Injection",
       y = "% Raw MS1 Area of Peptide\nNot Correctly Propionylated", 
       fill = "Batch") +
  theme_Publication() +
  theme(legend.box = "vertical",
        axis.text.x = element_text(angle = 45, vjust = 0.75, hjust=0.5))
dev.off()

pdf("propionylation_efficiency_summary.pdf",
    width = 5,
    height = 5)
ggplot(subset(df, !grepl("round", variable) & (replicate == 4 | replicate == 5)), aes(x = variable, y = value)) +
   geom_boxplot(aes(fill = replicate), outlier.shape = NA, position = position_dodge(width = 0.75)) +
  geom_jitter(aes(fill = replicate, shape = as.factor(multiplier)), 
              position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75, jitter.height = 0), 
              alpha = 0.75, size = 1.5) +
  scale_fill_manual(values = brewer.pal(5, "Set1")[4:5]) + 
  scale_y_continuous(labels = label_percent(accuracy = 1)) +
  scale_x_discrete(labels = c("Under\nPropionylation", "Over\nPropionlyation")) +
  labs(x = "", 
       shape = "# Cells Per Injection",
       y = "% Raw MS1 Area of Peptide\nNot Correctly Propionylated", 
       fill = "Batch") +
  theme_Publication() +
  theme(legend.box = "vertical")
dev.off()
```

### Correlation to abundance
- Propionlyation efficiency was estimated using H31-K9[un];K14[un]
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/filtered")

# correlation of abundance and prop eff
df <- assay(obj, "MS1") %>% 
  rownames_to_column("id") %>%
  pivot_longer(
    cols = -id, # This keeps the 'id' column fixed
    names_to = "variable",
    values_to = "value") %>%
  merge(as.data.frame(colData(obj)), by.x = "variable", by.y = "row.names") %>%
  filter(id == "H31-K9[un];K14[un]")

pdf("MS1.log1p_H31-K9[un];K14[un]_vs_prop.eff.underprop.summary.pdf",
    width = 5,
    height = 5)
ggplot(df, aes(x = log1p(value), y = prop.eff.underprop.summary)) +
  geom_point(aes(color = replicate), size = 1.5, alpha = 0.75) +
  stat_smooth(method = "lm") +
  stat_cor() +
  labs(x = "H31-K9[un];K14[un]\n  Raw Log1p Intensity",
       y = "% Raw MS1 Area of Peptide\nUnder Propionylated",
       color = "Batch") +
  scale_y_continuous(labels = label_percent(accuracy = 1)) +
  theme_Publication() +
  theme(legend.position = "right",       # Position legend on the right
        legend.direction = "vertical")   # Arrange legend items vertically
dev.off()

pdf("MS1.log1p_H31-K9[un];K14[un]_vs_prop.eff.overprop.pdf",
    width = 5,
    height = 5)
ggplot(df, aes(x = log1p(value), y = prop.eff.overprop)) +
  geom_point(aes(color = replicate), size = 1.5, alpha = 0.75) +
  stat_smooth(method = "lm") +
  stat_cor() +
  labs(x = "H31-K9[un];K14[un]\n  Raw Log1p Intensity",
       y = "% Raw MS1 Area of Peptide\nOver Propionylated",
       color = "Batch") +
  scale_y_continuous(labels = label_percent(accuracy = 1)) +
  theme_Publication() +
  theme(legend.position = "right",       # Position legend on the right
        legend.direction = "vertical")   # Arrange legend items vertically
dev.off()
```

## Intensity distribution

### MS1
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/filtered")

# transform
assays(obj.filter)[["MS1.log2"]] <- log2(assay(obj.filter, "MS1") + 1)

# imputation
assays(obj.filter)[["MS1.log2.impute"]] <- missForest(assay(obj.filter, "MS1.log2"))$ximp

# plot by cell
df <- assay(obj.filter, "MS1.log2") %>% 
  rownames_to_column("id") %>%
  pivot_longer(
    cols = -id, # This keeps the 'id' column fixed
    names_to = "variable",
    values_to = "value") %>%
  merge(as.data.frame(colData(obj.filter)), by.x = "variable", by.y = "row.names")

pdf("MS1.log2_distribution_filtered_cell.pdf",
    width = 8,
    height = 5)
ggplot(df, aes(x = as.factor(run_id), y = value, fill = experiment_replicate)) +
  geom_boxplot(width = 0.7) +
  theme_Publication() +
  labs(x = "Run", y = "Log2 MS1 intensity", fill = "Batch") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 10)) +
  facet_wrap(.~as.factor(multiplier), scales = "free_x", nrow = 3) +
  scale_fill_manual(values = c("auto_1" =  brewer.pal(5, "Set1")[1],
                              "auto_2" =  brewer.pal(5, "Set1")[2],
                              "auto_3" =  brewer.pal(5, "Set1")[3],
                              "nabut_1" =  brewer.pal(5, "Set1")[4],
                              "nabut_2" =  brewer.pal(5, "Set1")[5]),
                   labels = c("1",
                              "2",
                              "3",
                              "4",
                              "5"))
dev.off()

# plot by peptide
pdf("MS1.log2_distribution_filtered_peptide.pdf",
    width = 15,
    height = 10)
ggplot(df, aes(x = id, y = value)) +
  geom_boxplot() +
  theme_Publication() +
  labs(title = "Cell #", x = "Peptide", y = "Log2(MS1 intensity)", fill = "Batch") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 10)) +
  facet_wrap(.~as.factor(multiplier), scales = "free_x")
dev.off()

# heatmap
column_ha = HeatmapAnnotation(replicate = as.factor(obj.filter$replicate),
                              cell_number = as.factor(obj.filter$multiplier),
                              experiment = obj.filter$experiment,
                              run_id = obj.filter$run_id)
row_ha = rowAnnotation(charge = anno_barplot(rowData(obj.filter)$Precursor.Charge))

pdf("MS1.log2_filtered_heatmap.pdf",
    width = 10,
    height = 10)
set.seed(123)
Heatmap(scale(assay(obj.filter, "MS1.log2"), center = FALSE, scale = FALSE),
        top_annotation = column_ha,
        right_annotation = row_ha,
        heatmap_legend_param = list(title = "Log2(MS1 area)"),
        cluster_rows = TRUE,
        cluster_columns = TRUE,
        show_column_names = FALSE,
        row_names_gp = gpar(fontsize = 10))
dev.off()

# PCA
df <- colData(obj.filter)
df$multiplier <- factor(df$multiplier)
p <- pca(na.omit(assay(obj.filter, "MS1.log2")), 
         metadata = df,
         center = TRUE, 
         scale = TRUE)

screeplot(p)

pdf("MS1.log2_filtered_eigencor.pdf",
    width = 5,
    height = 5)
eigencorplot(p,
             main = "Principle component correlations",
             cexMain = 1.5,
             metavars = c("replicate", "date", "run_id", "multiplier", "total.ms1"),
             col = viridis(100),
             colCorval = 'firebrick',
             fontCorval = 2,
             cexCorval = 0.5,
             rotLabX = 45,
             posColKey = 'top')
dev.off()

pdf("MS1.log2_filtered_replicate_pca.pdf",
    width = 4,
    height = 5)
biplot(p,
       colby = "replicate", 
       shape = "multiplier",
       legendPosition = 'top',
       hline = 0, 
       vline = 0,
       lab = NULL) +
  labs(color = "Replicate",
       shape = "# Cells") +
  theme_Publication() +
  theme(legend.box="vertical", 
        legend.margin=margin())
dev.off()
```

### MS2
```{r}
# transform
assays(obj.filter)[["MS2.log2"]] <- log2(assay(obj.filter, "MS2") + 1)

# imputation
assays(obj.filter)[["MS2.log2.impute"]] <- missForest(assay(obj.filter, "MS2.log2"))$ximp

# plot by cell
df <- assay(obj.filter, "MS2.log2") %>% 
  rownames_to_column("id") %>%
  pivot_longer(
    cols = -id, # This keeps the 'id' column fixed
    names_to = "variable",
    values_to = "value") %>%
  merge(as.data.frame(colData(obj.filter)), by.x = "variable", by.y = "row.names")

pdf("MS2.log2_distribution_filtered_cell.pdf",
    width = 8,
    height = 10)
ggplot(df, aes(x = as.factor(run_id), y = value, fill = as.factor(replicate))) +
  geom_boxplot(width = 0.7) +
  theme_Publication() +
  labs(title = "Cell #", x = "Run", y = "Log2(MS2 intensity)", fill = "Batch") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 10)) +
  facet_wrap(.~as.factor(multiplier), scales = "free_x", nrow = 3)
dev.off()

# plot by peptide
pdf("MS2.log2_distribution_filtered_peptide.pdf",
    width = 15,
    height = 10)
ggplot(df, aes(x = id, y = value)) +
  geom_boxplot() +
  theme_Publication() +
  labs(title = "Cell #", x = "Peptide", y = "Log2(MS2 intensity)", fill = "Batch") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 10)) +
  facet_wrap(.~as.factor(multiplier), scales = "free_x")
dev.off()

# heatmap
# column_ha = HeatmapAnnotation(replicate = as.factor(obj.filter$replicate),
#                               cell_number = as.factor(obj.filter$multiplier),
#                               experiment = obj.filter$experiment,
#                               run_id = obj.filter$run_id)
# row_ha = rowAnnotation(charge = anno_barplot(rowData(obj.filter)$Precursor.Charge))
# 
# pdf("MS2.log2_filtered_heatmap.pdf",
#     width = 10,
#     height = 10)
# set.seed(123)
# Heatmap(scale(assay(obj.filter, "MS2.log2"), center = FALSE, scale = FALSE),
#         top_annotation = column_ha,
#         right_annotation = row_ha,
#         heatmap_legend_param = list(title = "Log2(MS2 area)"),
#         cluster_rows = TRUE,
#         cluster_columns = TRUE,
#         show_column_names = FALSE,
#         row_names_gp = gpar(fontsize = 10))
# dev.off()

# PCA
df <- colData(obj.filter)
df$multiplier <- factor(df$multiplier)
data <- assay(obj.filter, "MS2.log2")
data[data == 0] <- NA
p <- pca(na.omit(data),
         metadata = df,
         center = TRUE, 
         scale = TRUE)

screeplot(p)

pdf("MS2.log2_filtered_eigencor.pdf",
    width = 5,
    height = 5)
eigencorplot(p,
             main = "Principle component correlations",
             cexMain = 1.5,
             metavars = c("replicate", "date", "run_id", "multiplier", "total.ms2"),
             col = viridis(100),
             colCorval = 'firebrick',
             fontCorval = 2,
             cexCorval = 0.5,
             rotLabX = 45,
             posColKey = 'top')
dev.off()

pdf("MS2.log2_filtered_replicate_pca.pdf",
    width = 4,
    height = 5)
biplot(p,
       colby = "replicate", 
       shape = "multiplier",
       legendPosition = 'top',
       hline = 0, 
       vline = 0,
       lab = NULL) +
  labs(color = "Replicate",
       shape = "# Cells") +
  theme_Publication() +
  theme(legend.box="vertical", 
        legend.margin=margin())
dev.off()
```

### MS1 vs MS2
```{r}
df.1 <- assay(obj.filter, "MS1.log2.impute") %>% 
  rownames_to_column("id") %>%
  pivot_longer(
    cols = -id, # This keeps the 'id' column fixed
    names_to = "variable",
    values_to = "value")

df.2 <- assay(obj.filter, "MS2.log2.impute") %>% 
  rownames_to_column("id") %>%
  pivot_longer(
    cols = -id, # This keeps the 'id' column fixed
    names_to = "variable",
    values_to = "value")

df <- cbind(df.1, df.2$value)
colnames(df) <- c("id", "variable", "ms1", "ms2")
df <- merge(df, as.data.frame(colData(obj.filter)), by.x = "variable", by.y = "row.names")
df$id <- gsub(";", " ", df$id)

# by cell amount
pdf("MS1_vs_MS2_filtered_type.pdf",
    width = 8,
    height = 8)
ggplot(subset(df, multiplier != "0"), aes(x = ms1, y = ms2, color = as.factor(replicate))) +
  geom_point() +
  stat_smooth(method = "lm") +
  stat_cor() +
  facet_wrap(.~as.factor(multiplier)) +
  labs(title = "", x = "Log2(MS1)", y = "Log2(MS2)") +
  theme_Publication() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_color_discrete(name = "Replicate")
dev.off()

# just single cells
pdf("MS1_vs_MS2_filtered_1cell.pdf",
    width = 5,
    height = 5)
ggplot(subset(df, multiplier == "1"), aes(x = ms1, y = ms2, color = as.factor(replicate))) +
  geom_point() +
  stat_smooth(method = "lm") +
  stat_cor() +
  labs(title = "", x = "Log2(MS1)", y = "Log2(MS2)", color = "Replicate") +
  theme_Publication() +
  guides(color = guide_legend(override.aes = list(shape = 19)))
dev.off()

# facet peptides for single cells
pdf("MS1_vs_MS2_filtered_peptide.pdf",
    width = 25,
    height = 25)
ggplot(subset(df, multiplier == "1"), aes(x = ms1, y = ms2, color = as.factor(replicate))) +
  geom_point() +
  stat_smooth(method = "lm") +
  stat_cor() +
  facet_wrap(.~as.factor(multiplier)) +
  labs(title = "HIstone Peptides", x = "Log2(MS1)", y = "Log2(MS2)") +
  theme_Publication(base_size = 12) +
  scale_y_continuous(labels = function(x) format(x, scientific = TRUE)) +
  facet_wrap(.~id, scales = "free", labeller = labeller(id = label_wrap_gen(20))) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_color_discrete(name = "Replicate")
dev.off()

# by single cell averages
df.summary <- df %>%
  filter(multiplier == "1") %>%
  group_by(id, replicate) %>%
  summarise(ms1.mean = mean(ms1),
            ms2.mean = mean(ms2))

pdf("MS1_vs_MS2_filtered_peptide_1cell_mean.pdf",
    width = 5,
    height = 5)
ggplot(df.summary, aes(x = ms1.mean, y = ms2.mean, color = as.factor(replicate))) +
  geom_point() +
  stat_smooth(method = "lm") +
  stat_cor() +
  labs(title = "", x = "Log2(MS1)", y = "Log2(MS2)", color = "Replicate") +
  theme_Publication() +
  guides(color = guide_legend(override.aes = aes(label = ""))) +
  geom_text_repel(data = subset(df.summary, ms1.mean > 15 & ms2.mean < 12 & replicate == 1), 
            aes(label = id), max.overlaps	= Inf, color = "black", min.segment.length = 0, alpha = 0.75)
dev.off()
```

# Normalization
- peptide ratio normalization method
- group all histones ptms from same peptide
- filter out peptides which only have a single unmod/ptm
- calculate ratio of each ptm to all of ptms with same peptide

### MS1 peptidoform ratio
```{r}
dir.create("~/single-cell-histone-ptm/4-normalization/auto+nabut/normalization")
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/normalization")

ptm.ratio.ms1 <- calculate_ptm_ratio(obj.filter, assay = "MS1")

pdf("MS1.ratio_cell.pdf",
    width = 8,
    height = 8)
ggplot(ptm.ratio.ms1, aes(x = as.factor(run_id), y = ratio, fill = as.factor(experiment))) +
  geom_boxplot(width = 0.7) +
  theme_Publication() +
  labs(title = "Cell #", x = "Run", y = "MS1 peptide ratio", fill = "Experiment") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 10)) +
  facet_wrap(.~as.factor(multiplier), scales = "free_x", nrow = 2)
dev.off()

pdf("MS1.ratio.log1p_cell.pdf",
    width = 8,
    height = 8)
ggplot(ptm.ratio.ms1, aes(x = as.factor(run_id), y = ratio.log1p, fill = as.factor(experiment))) +
  geom_boxplot(width = 0.7) +
  theme_Publication() +
  labs(title = "Cell #", x = "Run", y = "MS1 peptide ratio", fill = "Experiment") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 10)) +
  facet_wrap(.~as.factor(multiplier), scales = "free_x", nrow = 2)
dev.off()
```

### MS2 peptidoform ratio
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/normalization")

ptm.ratio.ms2 <- calculate_ptm_ratio(obj.filter, assay = "MS2")

pdf("MS2.ratio_cell.pdf",
    width = 8,
    height = 8)
ggplot(ptm.ratio.ms2, aes(x = as.factor(run_id), y = ratio, fill = as.factor(experiment))) +
  geom_boxplot(width = 0.7) +
  theme_Publication() +
  labs(title = "Cell #", x = "Run", y = "MS2 peptide ratio", fill = "Experiment") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 10)) +
  facet_wrap(.~as.factor(multiplier), scales = "free_x", nrow = 2)
dev.off()

pdf("MS2.ratio.log1p_cell.pdf",
    width = 8,
    height = 8)
ggplot(ptm.ratio.ms2, aes(x = as.factor(run_id), y = ratio.log1p, fill = as.factor(experiment))) +
  geom_boxplot(width = 0.7) +
  theme_Publication() +
  labs(title = "Cell #", x = "Run", y = "MS2 peptide ratio", fill = "Experiment") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 10)) +
  facet_wrap(.~as.factor(multiplier), scales = "free_x", nrow = 2)
dev.off()
```

### Single PTM ratio
- summing percentages of same types of mods on same amino acid across peptides
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/normalization")

sptm.ratio.ms1 <- calculate_single_ptm_ratio(ptm.ratio.ms1, input = "long")
sptm.ratio.ms1.log1p <- log1p(sptm.ratio.ms1)
```

### Global PTM ratio
- Summing percentages of MS1
- Improvement: Use raw MS1 values to calculate new percentages
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/normalization")

global.ratio.ms1 <- calculate_global_ptm_ratio(sptm.ratio.ms1)
global.ratio.ms1.log1p <- log1p(global.ratio.ms1)
```

### Summarized H4 peptidoform ratio
- Summing percentages of MS1 for each combo of peptides: mono, di, tri, tetra
- Improvement: Use raw MS1 values to calculate new percentages
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/normalization")

h4.ac.ratio.ms1 <- calculate_h4_ac_summary_ratio(ptm.ratio.ms1, input = "long")
h4.ac.ratio.ms1.log1p <- log1p(h4.ac.ratio.ms1)
```

### Create new multi assay experiment
- Must create new one since number of peptides is different than previous object
- Using multi assay experiment to hold the summarized ratios
- IMPROVEMENT: add sample map
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/normalization")

# create summarized experiment from ratios
obj.filter.ratio <- load_summarized_experiment_ratio(ptm.ratio.ms1, 
                                                     ptm.ratio.ms2, 
                                                     obj.filter)

# create new replicate IDs
obj.filter.ratio$experiment_replicate <- paste(obj.filter.ratio$experiment,
                                               obj.filter.ratio$replicate,
                                               sep = "_")

# create sample map for adding in other matrices
# map <- data.frame(
#     primary = colnames(obj.filter.ratio),
#     colnames = colnames(obj.filter.ratio),
#     stringsAsFactors = FALSE)
# 
# maplist <- list(ptm.ratio = map, 
#                 sptm.ratio = map,
#                 global.ratio = map,
#                 h4.ac.ratio = map)
# 
# sampMap <- listToMap(maplist)

# create multi assay experiment that contains multiple summarized experiments 
obj.filter.ratio.multi <- MultiAssayExperiment(experiments = ExperimentList(list(ptm.ratio = obj.filter.ratio,
                                              sptm.ratio = SummarizedExperiment(list(MS1.ratio = sptm.ratio.ms1,
                                                                                     MS1.ratio.log1p = sptm.ratio.ms1.log1p)),
                                              global.ratio = SummarizedExperiment(list(MS1.ratio = global.ratio.ms1,
                                                                                       MS1.ratio.log1p = global.ratio.ms1.log1p)),
                                              h4.ac.ratio = SummarizedExperiment(list(MS1.ratio = h4.ac.ratio.ms1,
                                                                                      MS1.ratio.log1p = h4.ac.ratio.ms1.log1p)))),
                             colData = colData(obj.filter.ratio))

# add unique replicate ids
obj.filter.ratio.multi$replicate.unique <- factor(as.numeric(factor(obj.filter.ratio.multi$experiment_replicate)))

# make peptide notes readable
rowData(obj.filter.ratio.multi[["ptm.ratio"]])$PTM.Name.Readable <- ptm_readable_translation(rownames(rowData(obj.filter.ratio.multi[["ptm.ratio"]])))
```

### Visualize

#### MS1 ratio
- 100 cells removed from PCA
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/normalization")

# plot by cell
df <- assay(obj.filter.ratio.multi[["ptm.ratio"]], "MS1.ratio.log1p") %>% 
  rownames_to_column("id") %>%
  pivot_longer(
    cols = -id, # This keeps the 'id' column fixed
    names_to = "variable",
    values_to = "value") %>%
  merge(as.data.frame(colData(obj.filter)), by.x = "variable", by.y = "row.names")

pdf("MS1.log2_distribution_filtered_cell.pdf",
    width = 8,
    height = 10)
ggplot(df, aes(x = as.factor(run_id), y = value, fill = as.factor(replicate))) +
  geom_boxplot(width = 0.7) +
  theme_Publication() +
  labs(title = "Cell #", x = "Run", y = "Log2(MS1 intensity)", fill = "Batch") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 10)) +
  facet_wrap(.~as.factor(multiplier), scales = "free_x", nrow = 3)
dev.off()

# heatmap
column_ha = HeatmapAnnotation(replicate = as.factor(obj.filter.ratio.multi$replicate),
                              experiment = obj.filter.ratio.multi$experiment,
                              cell_number = as.factor(obj.filter.ratio.multi$multiplier),
                              condition = obj.filter.ratio.multi$condition,
                              run_id = obj.filter.ratio.multi$run_id)

pdf("MS1.ratio.log1p_heatmap.pdf",
    width = 20,
    height = 20)
set.seed(123)
Heatmap(assay(experiments(obj.filter.ratio.multi)[["ptm.ratio"]], "MS1.ratio.log1p"),
        top_annotation = column_ha,
        #right_annotation = row_ha,
        heatmap_legend_param = list(title = "MS1 Peptide Ratio"),
        cluster_rows = TRUE,
        cluster_columns = TRUE,
        show_column_names = TRUE,
        row_names_gp = gpar(fontsize = 10))
dev.off()

# PCA
obj.filter.ratio.multi.1cell <- obj.filter.ratio.multi[,obj.filter.ratio.multi$multiplier == 1]

p <- pca(assay(experiments(obj.filter.ratio.multi.1cell)[["ptm.ratio"]], "MS1.ratio.log1p"),
         metadata = colData(obj.filter.ratio.multi.1cell),
         center = TRUE, 
         scale = TRUE)

pdf("MS1.ratio.log1p_screeplot.pdf",
    width = 4,
    height = 4)
screeplot(p,
          components = getComponents(p)[1:5])
dev.off()

pdf("MS1.ratio.log1p_eigencor.pdf",
    width = 6,
    height = 4)
eigencorplot(p,
             components = getComponents(p, seq_len(5)),
             main = "Principle component correlations",
             cexMain = 1.5,
             metavars = c("experiment_replicate", 
                          "total.ms1", 
                          "prop.eff.underprop.summary", 
                          "prop.eff.overprop"),
             col = viridis(100),
             colCorval = 'firebrick',
             fontCorval = 2,
             cexCorval = 0.5,
             rotLabX = 45,
             posColKey = 'top')
dev.off()



pdf("MS1.ratio.log1p_experiment_pca.pdf",
    width = 5,
    height = 5)
biplot(p,
       colby = "experiment", 
       shape = "multiplier",
       legendPosition = 'top',
       hline = 0, 
       vline = 0,
       lab = NULL) +
  labs(color = "Experiment",
       shape = "Multiplier") +
  theme_Publication() +
  theme(legend.box="vertical", 
        legend.margin=margin())
dev.off()

pdf("MS1.ratio.log1p_replicate_pca.pdf",
    width = 5,
    height = 5)
biplot(p,
       colby = "experiment_replicate", 
       shape = "condition",
       legendPosition = 'top',
       hline = 0, 
       vline = 0,
       lab = NULL) +
  labs(color = "Batch",
       shape = "Condition") +
  theme_Publication() +
  theme(legend.box="vertical",
        legend.margin=margin()) +
  scale_color_manual(values = c("auto_1" =  brewer.pal(5, "Set1")[1],
                                "auto_2" =  brewer.pal(5, "Set1")[2],
                                "auto_3" =  brewer.pal(5, "Set1")[3],
                                "nabut_1" =  brewer.pal(5, "Set1")[4],
                                "nabut_2" =  brewer.pal(5, "Set1")[5]),
                     labels = c("1",
                                "2",
                                "3",
                                "4",
                                "5"))
dev.off()

pdf("MS1.ratio.log1p_condition_pca.pdf",
    width = 6,
    height = 5)
biplot(p,
       colby = "condition", 
       shape = "multiplier",
       legendPosition = 'top',
       hline = 0, 
       vline = 0,
       lab = NULL) +
  labs(color = "Condition",
       shape = "Multiplier") +
  theme_Publication() +
  theme(legend.box="vertical", 
        legend.margin=margin())
dev.off()
```

#### MS2 ratio
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/normalization")

# heatmap
column_ha = HeatmapAnnotation(replicate = as.factor(obj.filter.ratio$replicate),
                              experiment = obj.filter.ratio$experiment,
                              cell_number = as.factor(obj.filter.ratio$multiplier),
                              condition = obj.filter.ratio$condition,
                              run_id = obj.filter.ratio$run_id)

pdf("MS2.ratio.log1p_heatmap.pdf",
    width = 10,
    height = 10)
set.seed(123)
Heatmap(assay(experiments(obj.filter.ratio.multi)[["ptm.ratio"]], "MS2.ratio.log1p"),
        top_annotation = column_ha,
        #right_annotation = row_ha,
        heatmap_legend_param = list(title = "MS2 Peptide Ratio"),
        cluster_rows = TRUE,
        cluster_columns = TRUE,
        show_column_names = FALSE,
        row_names_gp = gpar(fontsize = 10))
dev.off()

# PCA
p <- pca(na.omit(assay(experiments(obj.filter.ratio.multi)[["ptm.ratio"]], "MS2.ratio.log1p")),
         metadata = colData(obj.filter.ratio),
         center = TRUE, 
         scale = TRUE)

pdf("MS2.ratio.log1p_screeplot.pdf",
    width = 5,
    height = 5)
screeplot(p,
          components = getComponents(p)[1:10])
dev.off()

# pdf("MS2.ratio.log1p_eigencor.pdf",
#     width = 5,
#     height = 5)
# eigencorplot(p,
#              main = "Principle component correlations",
#              cexMain = 1.5,
#             metavars = c("experiment_replicate", 
#                           "multiplier", 
#                           "total.ms2", 
#                           "prop.eff.underprop.summary", 
#                           "prop.eff.overprop"),
#              col = viridis(100),
#              colCorval = 'firebrick',
#              fontCorval = 2,
#              cexCorval = 0.5,
#              rotLabX = 45,
#              posColKey = 'top')
# dev.off()

pdf("MS2.ratio.log1p_experiment_pca.pdf",
    width = 5,
    height = 5)
biplot(p,
       colby = "experiment", 
       shape = "multiplier",
       legendPosition = 'top',
       hline = 0, 
       vline = 0,
       lab = NULL) +
  labs(color = "Experiment",
       shape = "Multiplier") +
  theme_Publication() +
  theme(legend.box="vertical", 
        legend.margin=margin())
dev.off()

pdf("MS2.ratio.log1p_replicate_pca.pdf",
    width = 5,
    height = 5)
biplot(p,
       colby = "replicate", 
       shape = "multiplier",
       legendPosition = 'top',
       hline = 0, 
       vline = 0,
       lab = NULL) +
  labs(color = "Replicate",
       shape = "Multiplier") +
  theme_Publication() +
  theme(legend.box="vertical", 
        legend.margin=margin())
dev.off()

pdf("MS2.ratio.log1p_condition_pca.pdf",
    width = 5,
    height = 5)
biplot(p,
       colby = "condition", 
       shape = "multiplier",
       legendPosition = 'top',
       hline = 0, 
       vline = 0,
       lab = NULL) +
  labs(color = "Condition",
       shape = "Multiplier") +
  theme_Publication() +
  theme(legend.box="vertical", 
        legend.margin=margin())
dev.off()
```

#### Single PTM ratio
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/normalization")

# heatmap
column_ha = HeatmapAnnotation(replicate = as.factor(obj.filter.ratio.multi$replicate),
                              experiment = obj.filter.ratio.multi$experiment,
                              cell_number = as.factor(obj.filter.ratio.multi$multiplier),
                              condition = obj.filter.ratio.multi$condition,
                              run_id = obj.filter.ratio.multi$run_id)

pdf("MS1.ratio.single_heatmap.pdf",
    width = 10,
    height = 10)
set.seed(123)
Heatmap(assay(experiments(obj.filter.ratio.multi)[["sptm.ratio"]], "MS1.ratio"),
        top_annotation = column_ha,
        #right_annotation = row_ha,
        heatmap_legend_param = list(title = "PTM relative abundance"),
        cluster_rows = TRUE,
        cluster_columns = TRUE,
        show_column_names = FALSE,
        row_names_gp = gpar(fontsize = 10))
dev.off()

pdf("MS1.ratio.log1p.single_heatmap.pdf",
    width = 10,
    height = 10)
set.seed(123)
Heatmap(assay(experiments(obj.filter.ratio.multi)[["sptm.ratio"]], "MS1.ratio.log1p"),
        top_annotation = column_ha,
        #right_annotation = row_ha,
        heatmap_legend_param = list(title = "PTM relative abundance"),
        cluster_rows = TRUE,
        cluster_columns = TRUE,
        show_column_names = FALSE,
        row_names_gp = gpar(fontsize = 10))
dev.off()
```

#### Global PTM ratio
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/normalization")

# heatmap
column_ha = HeatmapAnnotation(replicate = as.factor(obj.filter.ratio.multi$replicate),
                              experiment = obj.filter.ratio.multi$experiment,
                              cell_number = as.factor(obj.filter.ratio.multi$multiplier),
                              condition = obj.filter.ratio.multi$condition,
                              run_id = obj.filter.ratio.multi$run_id)

pdf("MS1.ratio.global_heatmap.pdf",
    width = 10,
    height = 3)
set.seed(123)
Heatmap(assay(experiments(obj.filter.ratio.multi)[["global.ratio"]], "MS1.ratio"),
        top_annotation = column_ha,
        #right_annotation = row_ha,
        heatmap_legend_param = list(title = "PTM relative abundance"),
        cluster_rows = TRUE,
        cluster_columns = TRUE,
        show_column_names = FALSE,
        row_names_gp = gpar(fontsize = 10))
dev.off()

pdf("MS1.ratio.log1p.global_heatmap.pdf",
    width = 10,
    height = 3)
set.seed(123)
Heatmap(assay(experiments(obj.filter.ratio.multi)[["global.ratio"]], "MS1.ratio.log1p"),
        top_annotation = column_ha,
        #right_annotation = row_ha,
        heatmap_legend_param = list(title = "PTM relative abundance"),
        cluster_rows = TRUE,
        cluster_columns = TRUE,
        show_column_names = FALSE,
        row_names_gp = gpar(fontsize = 10))
dev.off()
```

#### Summarized H4 PTM ratio
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/normalization")

# heatmap
column_ha = HeatmapAnnotation(replicate = as.factor(obj.filter.ratio.multi$replicate),
                              experiment = obj.filter.ratio.multi$experiment,
                              cell_number = as.factor(obj.filter.ratio.multi$multiplier),
                              condition = obj.filter.ratio.multi$condition,
                              run_id = obj.filter.ratio.multi$run_id)

pdf("MS1.ratio.h4_heatmap.pdf",
    width = 10,
    height = 3)
set.seed(123)
Heatmap(assay(experiments(obj.filter.ratio.multi)[["h4.ac.ratio"]], "MS1.ratio"),
        top_annotation = column_ha,
        #right_annotation = row_ha,
        heatmap_legend_param = list(title = "PTM relative abundance"),
        cluster_rows = TRUE,
        cluster_columns = TRUE,
        show_column_names = FALSE,
        row_names_gp = gpar(fontsize = 10))
dev.off()

pdf("MS1.ratio.log1p.h4_heatmap.pdf",
    width = 10,
    height = 3)
set.seed(123)
Heatmap(assay(experiments(obj.filter.ratio.multi)[["h4.ac.ratio"]], "MS1.ratio.log1p"),
        top_annotation = column_ha,
        #right_annotation = row_ha,
        heatmap_legend_param = list(title = "PTM relative abundance"),
        cluster_rows = TRUE,
        cluster_columns = TRUE,
        show_column_names = FALSE,
        row_names_gp = gpar(fontsize = 10))
dev.off()
```

#### Propionylation efficiency correlates
- Propionlyation efficiency was estimated using H31-K18[un];K23[un]
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/normalization")

# correlation of abundance and prop eff
df <- assay(experiments(obj.filter.ratio.multi)[["ptm.ratio"]], "MS1.ratio.log1p") %>% 
  rownames_to_column("id") %>%
  pivot_longer(
    cols = -id, # This keeps the 'id' column fixed
    names_to = "variable",
    values_to = "value") %>%
  merge(as.data.frame(colData(obj.filter.ratio.multi)), by.x = "variable", by.y = "row.names") %>%
  filter(id == "H31-K18[un];K23[un]")

pdf("MS1.ratio.log1p_vs_prop.eff.underprop.summary.pdf",
    width = 5,
    height = 5)
ggplot(df, aes(x = value, y = prop.eff.underprop.summary)) +
  geom_point(aes(color = experiment_replicate), size = 1.5, alpha = 0.75) +
  stat_smooth(method = "lm") +
  stat_cor() +
  labs(x = "H31-K18[un];K23[un]\nRelative Abundance",
       y = "% Raw MS1 Area of Peptide\nNot Correctly Propionylated",
       color = "Batch") +
  scale_y_continuous(labels = label_percent(accuracy = 1)) +
  theme_Publication() +
  theme(legend.position = "right",       # Position legend on the right
        legend.direction = "vertical")   # Arrange legend items vertically
dev.off()
```

### Save object
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/")

saveRDS(obj.filter.ratio.multi, "obj.filter.ratio.multi.RDS")
```

### Output tables
```{r}
dir.create("~/single-cell-histone-ptm/4-normalization/auto+nabut/normalization")
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/normalization")

write.csv(assay(obj.filter.ratio.multi[["ptm.ratio"]], "RT"), "RT.csv")
write.csv(assay(obj.filter.ratio.multi[["ptm.ratio"]], "RT")/40, "RT.relative.csv")
write.csv(rowData(obj.filter.ratio.multi[["ptm.ratio"]])[,c(1:5)], "peptidoform-annotation.csv")

# untransformed
write.csv(assay(obj.filter.ratio.multi[["ptm.ratio"]], "MS1.ratio"), "MS1.ratio.csv")
write.csv(assay(obj.filter.ratio.multi[["sptm.ratio"]], "MS1.ratio"), "MS1.ratio.single.csv")
write.csv(assay(obj.filter.ratio.multi[["global.ratio"]], "MS1.ratio"), "MS1.ratio.global.csv")
write.csv(assay(obj.filter.ratio.multi[["h4.ac.ratio"]], "MS1.ratio"), "MS1.ratio.h4.csv")

# transformed
write.csv(assay(obj.filter.ratio.multi[["ptm.ratio"]], "MS1.ratio.log1p"), "MS1.ratio.log1p.csv")
write.csv(assay(obj.filter.ratio.multi[["sptm.ratio"]], "MS1.ratio.log1p"), "MS1.ratio.log1p.single.csv")
write.csv(assay(obj.filter.ratio.multi[["global.ratio"]], "MS1.ratio.log1p"), "MS1.ratio.log1p.global.csv")
write.csv(assay(obj.filter.ratio.multi[["h4.ac.ratio"]], "MS1.ratio.log1p"), "MS1.ratio.log1p.h4.csv")
```

# Batch correction

## MS1
```{r}
dir.create("~/single-cell-histone-ptm/4-normalization/auto+nabut/batch-corrected")
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/batch-corrected")

assays(obj.filter.ratio.multi[["ptm.ratio"]], withDimnames=FALSE)[["MS1.ratio.log1p.batchCorrect"]] <- 
  removeBatchEffect(assay(obj.filter.ratio.multi[["ptm.ratio"]], "MS1.ratio.log1p"),
                    batch = obj.filter.ratio.multi$date,
                    group = obj.filter.ratio.multi$condition)
```

## MS2
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/batch-corrected")

assays(obj.filter.ratio.multi[["ptm.ratio"]], withDimnames=FALSE)[["MS2.ratio.batchCorrect"]] <- 
  removeBatchEffect(assay(obj.filter.ratio.multi[["ptm.ratio"]], "MS2.ratio"),
                    batch = obj.filter.ratio.multi$date,
                    group = obj.filter.ratio.multi$condition)
```

## Single PTM ratio
```{r}
assays(obj.filter.ratio.multi[["sptm.ratio"]], withDimnames=FALSE)[["MS1.ratio.log1p.batchCorrect"]] <- 
  removeBatchEffect(assay(obj.filter.ratio.multi[["sptm.ratio"]], "MS1.ratio.log1p"),
                    batch = obj.filter.ratio.multi$date,
                    group = obj.filter.ratio.multi$condition)
```

## Global PTM ratio
```{r}
assays(obj.filter.ratio.multi[["global.ratio"]], withDimnames=FALSE)[["MS1.ratio.log1p.batchCorrect"]] <- 
  removeBatchEffect(assay(obj.filter.ratio.multi[["global.ratio"]], "MS1.ratio.log1p"),
                    batch = obj.filter.ratio.multi$date,
                    group = obj.filter.ratio.multi$condition)
```

## Summarized H4 PTM ratio
```{r}
assays(obj.filter.ratio.multi[["h4.ac.ratio"]], withDimnames=FALSE)[["MS1.ratio.log1p.batchCorrect"]] <- 
  removeBatchEffect(assay(obj.filter.ratio.multi[["h4.ac.ratio"]], "MS1.ratio.log1p"),
                    batch = obj.filter.ratio.multi$date,
                    group = obj.filter.ratio.multi$condition)
```

## Visualize

### MS1
- 100 cells removed from PCA
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/batch-corrected")

# plot by cell
df <- assay(obj.filter.ratio.multi[["ptm.ratio"]], "MS1.ratio.log1p.batchCorrect") %>% 
  as.data.frame() %>%
  rownames_to_column("id") %>%
  pivot_longer(
    cols = -id, # This keeps the 'id' column fixed
    names_to = "variable",
    values_to = "value") %>%
  merge(as.data.frame(colData(obj.filter)), by.x = "variable", by.y = "row.names")

pdf("MS1.ratio.log1p.batchCorrect_distribution.pdf",
    width = 8,
    height = 5)
ggplot(df, aes(x = as.factor(run_id), y = value, fill = experiment_replicate)) +
  geom_boxplot(width = 0.7) +
  theme_Publication() +
  labs(x = "Run", y = "Log1p Relative Abundance", fill = "Batch") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 10)) +
  facet_wrap(.~as.factor(multiplier), scales = "free_x", nrow = 3) +
  scale_fill_manual(values = c("auto_1" =  brewer.pal(5, "Set1")[1],
                              "auto_2" =  brewer.pal(5, "Set1")[2],
                              "auto_3" =  brewer.pal(5, "Set1")[3],
                              "nabut_1" =  brewer.pal(5, "Set1")[4],
                              "nabut_2" =  brewer.pal(5, "Set1")[5]),
                   labels = c("1",
                              "2",
                              "3",
                              "4",
                              "5"))
dev.off()

# heatmap
column_ha = HeatmapAnnotation(replicate = as.factor(obj.filter.ratio.multi$replicate),
                              experiment = obj.filter.ratio.multi$experiment,
                              condition = obj.filter.ratio.multi$condition, 
                              cell_number = as.factor(obj.filter.ratio.multi$multiplier),
                              run_id = obj.filter.ratio.multi$run_id)
row_ha = rowAnnotation(charge = anno_barplot(rowData(obj.filter.ratio.multi[["ptm.ratio"]])$Precursor.Charge))

pdf("MS1.ratio.log1p.batchCorrect_heatmap.pdf",
    width = 10,
    height = 10)
set.seed(123)
Heatmap(assay(obj.filter.ratio.multi[["ptm.ratio"]], "MS1.ratio.log1p.batchCorrect"),
        top_annotation = column_ha,
        right_annotation = row_ha,
        heatmap_legend_param = list(title = "Normalized Peptide Ratio"),
        cluster_rows = TRUE,
        cluster_columns = TRUE,
        show_column_names = FALSE,
        row_names_gp = gpar(fontsize = 10))
dev.off()

# PCA
obj.filter.ratio.multi.1cell <- obj.filter.ratio.multi[,obj.filter.ratio.multi$multiplier == 1]

p <- pca(assay(obj.filter.ratio.multi.1cell[["ptm.ratio"]], "MS1.ratio.log1p.batchCorrect"),
         metadata = colData(obj.filter.ratio.multi.1cell),
         center = TRUE, 
         scale = TRUE)

pdf("MS1.ratio.log1p_screeplot.pdf",
    width = 4,
    height = 4)
screeplot(p,
          components = getComponents(p)[1:5])
dev.off()

pdf("MS1.ratio.log1p.batchCorrect_eigencor.pdf",
    width = 4,
    height = 4)
eigencorplot(p,
             components = getComponents(p)[1:5],
             main = "Principle component correlations",
             cexMain = 1.5,
             metavars = c("experiment_replicate", "condition", "total.ms1"),
             col = viridis(100),
             colCorval = 'firebrick',
             fontCorval = 2,
             cexCorval = 0.5,
             rotLabX = 45,
             posColKey = 'top')
dev.off()

pdf("MS1.ratio.log1p.batchCorrect_experiment_pca.pdf",
    width = 5,
    height = 5)
biplot(p,
       colby = "experiment", 
       shape = "multiplier",
       legendPosition = 'top',
       hline = 0, 
       vline = 0,
       lab = NULL) +
  labs(color = "Experiment",
       shape = "Multiplier") +
  theme_Publication() +
  theme(legend.box="vertical", 
        legend.margin=margin())
dev.off()

pdf("MS1.ratio.log1p_replicate_pca.pdf",
    width = 5,
    height = 5)
biplot(p,
       colby = "experiment_replicate", 
       shape = "condition",
       legendPosition = 'top',
       hline = 0, 
       vline = 0,
       lab = NULL) +
  labs(color = "Batch",
       shape = "Condition") +
  theme_Publication() +
  theme(legend.box="vertical",
        legend.margin=margin()) +
  scale_color_manual(values = c("auto_1" =  brewer.pal(5, "Set1")[1],
                                "auto_2" =  brewer.pal(5, "Set1")[2],
                                "auto_3" =  brewer.pal(5, "Set1")[3],
                                "nabut_1" =  brewer.pal(5, "Set1")[4],
                                "nabut_2" =  brewer.pal(5, "Set1")[5]),
                     labels = c("1",
                                "2",
                                "3",
                                "4",
                                "5"))
dev.off()

pdf("MS1.ratio.log1p.batchCorrect_condition_pca.pdf",
    width = 5,
    height = 5)
biplot(p,
       colby = "condition", 
       shape = "multiplier",
       legendPosition = 'top',
       hline = 0, 
       vline = 0,
       lab = NULL) +
  labs(color = "Condition",
       shape = "Multiplier") +
  theme_Publication() +
  theme(legend.box="vertical", 
        legend.margin=margin())
dev.off()

pdf("MS1.ratio.log1p.batchCorrect_condition_labels_pca.pdf",
    width = 5,
    height = 5)
biplot(p,
       colby = "condition", 
       shape = "multiplier",
       legendPosition = 'top',
       hline = 0, 
       vline = 0) +
  labs(color = "Condition",
       shape = "Multiplier") +
  theme_Publication() +
  theme(legend.box="vertical", 
        legend.margin=margin())
dev.off()
```

### MS2
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/batch-corrected")

# heatmap
column_ha = HeatmapAnnotation(replicate = as.factor(obj.filter.ratio.multi$replicate),
                              experiment = obj.filter.ratio.multi$experiment,
                              condition = obj.filter.ratio.multi$condition, 
                              cell_number = as.factor(obj.filter.ratio.multi$multiplier),
                              run_id = obj.filter.ratio.multi$run_id)
row_ha = rowAnnotation(charge = anno_barplot(rowData(obj.filter.ratio.multi[["ptm.ratio"]])$Precursor.Charge))

pdf("MS2.ratio.batchCorrect_heatmap.pdf",
    width = 10,
    height = 10)
set.seed(123)
Heatmap(assay(obj.filter.ratio.multi[["ptm.ratio"]], "MS2.ratio.batchCorrect"),
        top_annotation = column_ha,
        right_annotation = row_ha,
        heatmap_legend_param = list(title = "Normalized Peptide Ratio"),
        cluster_rows = TRUE,
        cluster_columns = TRUE,
        show_column_names = FALSE,
        row_names_gp = gpar(fontsize = 10))
dev.off()

# PCA
p <- pca(na.omit(assay(obj.filter.ratio.multi[["ptm.ratio"]], "MS2.ratio.batchCorrect")),
         metadata = colData(obj.filter.ratio.multi),
         center = TRUE, 
         scale = TRUE)

pdf("MS2.ratio_screeplot.pdf",
    width = 5,
    height = 5)
screeplot(p,
          components = getComponents(p)[1:10])
dev.off()

# pdf("MS2.ratio.batchCorrect_eigencor.pdf",
#     width = 5,
#     height = 5)
# eigencorplot(p,
#              main = "Principle component correlations",
#              cexMain = 1.5,
#              metavars = c("experiment_replicate", "condition", "total.ms2"),
#              col = viridis(100),
#              colCorval = 'firebrick',
#              fontCorval = 2,
#              cexCorval = 0.5,
#              rotLabX = 45,
#              posColKey = 'top')
# dev.off()

pdf("MS2.ratio.batchCorrect_experiment_pca.pdf",
    width = 5,
    height = 5)
biplot(p,
       colby = "experiment", 
       shape = "multiplier",
       legendPosition = 'top',
       hline = 0, 
       vline = 0,
       lab = NULL) +
  labs(color = "Experiment",
       shape = "Multiplier") +
  theme_Publication() +
  theme(legend.box="vertical", 
        legend.margin=margin())
dev.off()

pdf("MS2.ratio.batchCorrect_replicate_pca.pdf",
    width = 5,
    height = 5)
biplot(p,
       colby = "replicate", 
       shape = "multiplier",
       legendPosition = 'top',
       hline = 0, 
       vline = 0,
       lab = NULL) +
  labs(color = "Replicate",
       shape = "Multiplier") +
  theme_Publication() +
  theme(legend.box="vertical", 
        legend.margin=margin())
dev.off()

pdf("MS2.ratio.batchCorrect_condition_pca.pdf",
    width = 5,
    height = 5)
biplot(p,
       colby = "condition", 
       shape = "multiplier",
       legendPosition = 'top',
       hline = 0, 
       vline = 0,
       lab = NULL) +
  labs(color = "Condition",
       shape = "Multiplier") +
  theme_Publication() +
  theme(legend.box="vertical", 
        legend.margin=margin())
dev.off()
```

### Single PTM ratio
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/batch-corrected")

# heatmap
column_ha = HeatmapAnnotation(replicate = as.factor(obj.filter.ratio.multi$replicate),
                              experiment = obj.filter.ratio.multi$experiment,
                              condition = obj.filter.ratio.multi$condition, 
                              cell_number = as.factor(obj.filter.ratio.multi$multiplier),
                              run_id = obj.filter.ratio.multi$run_id)

pdf("MS1.ratio.log1p.batchCorrect.single_heatmap.pdf",
    width = 10,
    height = 10)
set.seed(123)
Heatmap(assay(experiments(obj.filter.ratio.multi)[["sptm.ratio"]], "MS1.ratio.log1p.batchCorrect"),
        top_annotation = column_ha,
        heatmap_legend_param = list(title = "PTM relative abundance"),
        cluster_rows = TRUE,
        cluster_columns = TRUE,
        show_column_names = FALSE,
        row_names_gp = gpar(fontsize = 10))
dev.off()
```

### Global PTM ratio
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/batch-corrected")

# heatmap
column_ha = HeatmapAnnotation(replicate = as.factor(obj.filter.ratio.multi$replicate),
                              experiment = obj.filter.ratio.multi$experiment,
                              cell_number = as.factor(obj.filter.ratio.multi$multiplier),
                              condition = obj.filter.ratio.multi$condition,
                              run_id = obj.filter.ratio.multi$run_id)

pdf("MS1.ratio.log1p.batchCorrect.global_heatmap.pdf",
    width = 10,
    height = 3)
set.seed(123)
Heatmap(assay(experiments(obj.filter.ratio.multi)[["global.ratio"]], "MS1.ratio.log1p.batchCorrect"),
        top_annotation = column_ha,
        heatmap_legend_param = list(title = "PTM relative abundance"),
        cluster_rows = TRUE,
        cluster_columns = TRUE,
        show_column_names = FALSE,
        row_names_gp = gpar(fontsize = 10))
dev.off()
```

### Summarized H4 PTM ratio
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/batch-corrected")

# heatmap
column_ha = HeatmapAnnotation(replicate = as.factor(obj.filter.ratio.multi$replicate),
                              experiment = obj.filter.ratio.multi$experiment,
                              cell_number = as.factor(obj.filter.ratio.multi$multiplier),
                              condition = obj.filter.ratio.multi$condition,
                              run_id = obj.filter.ratio.multi$run_id)

pdf("MS1.ratio.log1p.batchCorrect.h4_heatmap.pdf",
    width = 10,
    height = 3)
set.seed(123)
Heatmap(assay(experiments(obj.filter.ratio.multi)[["h4.ac.ratio"]], "MS1.ratio.log1p.batchCorrect"),
        top_annotation = column_ha,
        heatmap_legend_param = list(title = "PTM relative abundance"),
        cluster_rows = TRUE,
        cluster_columns = TRUE,
        show_column_names = FALSE,
        row_names_gp = gpar(fontsize = 10))
dev.off()
```

### Propionylation efficiency correlates
- Propionlyation efficiency was estimated using H31-K18[un];K23[un]
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/batch-corrected")

# correlation of abundance and prop eff
df <- assay(obj.filter.ratio.multi[["ptm.ratio"]], "MS1.ratio.log1p.batchCorrect") %>% 
  as.data.frame() %>%
  rownames_to_column("id") %>%
  pivot_longer(
    cols = -id, # This keeps the 'id' column fixed
    names_to = "variable",
    values_to = "value") %>%
  merge(as.data.frame(colData(obj.filter.ratio.multi)), by.x = "variable", by.y = "row.names") %>%
  filter(id == "H31-K18[un];K23[un]")

pdf("MS1.ratio.log1p_vs_prop.eff.underprop.summary.pdf",
    width = 5,
    height = 5)
ggplot(df, aes(x = value, y = prop.eff.underprop.summary)) +
  geom_point(aes(color = experiment_replicate), size = 1.5, alpha = 0.75) +
  stat_smooth(method = "lm") +
  stat_cor() +
  labs(x = "H31-K18[un];K23[un]\nRelative Abundance",
       y = "% Raw MS1 Area of Peptide\nNot Correctly Propionylated",
       color = "Batch") +
  scale_y_continuous(labels = label_percent(accuracy = 1)) +
  theme_Publication() +
  theme(legend.position = "right",       # Position legend on the right
        legend.direction = "vertical")   # Arrange legend items vertically
dev.off()
```

## Save object
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut")

saveRDS(obj.filter.ratio.multi, "obj.filter.ratio.multi.RDS")
```

## Output tables
```{r}
setwd("~/single-cell-histone-ptm/4-normalization/auto+nabut/batch-corrected")

write.csv(assay(obj.filter.ratio.multi[["ptm.ratio"]], "RT"), "RT.csv")
write.csv(assay(obj.filter.ratio.multi[["ptm.ratio"]], "RT")/40, "RT.relative.csv")
write.csv(rowData(obj.filter.ratio.multi[["ptm.ratio"]])[,c(1:5)], "peptidoform-annotation.csv")

# batch corrected
write.csv(assay(obj.filter.ratio.multi[["ptm.ratio"]], "MS1.ratio.log1p.batchCorrect"), "MS1.ratio.log1p.batchCorrect.csv")
write.csv(assay(obj.filter.ratio.multi[["sptm.ratio"]], "MS1.ratio.log1p.batchCorrect"), "MS1.ratio.log1p.single.batchCorrect.csv")
write.csv(assay(obj.filter.ratio.multi[["global.ratio"]], "MS1.ratio.log1p.batchCorrect"), "MS1.ratio.log1p.global.batchCorrect.csv")
write.csv(assay(obj.filter.ratio.multi[["h4.ac.ratio"]], "MS1.ratio.log1p.batchCorrect"), "MS1.ratio.log1p.h4.batchCorrect.csv")
```